<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Multiple Comparisons | Worked Bioninformatics, Statistics, and Machine Learning Examples</title>
  <meta name="description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  <meta name="generator" content="bookdown 0.17 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Multiple Comparisons | Worked Bioninformatics, Statistics, and Machine Learning Examples" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://sw1.github.io/teaching/" />
  
  <meta property="og:description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  <meta name="github-repo" content="sw1/teaching" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Multiple Comparisons | Worked Bioninformatics, Statistics, and Machine Learning Examples" />
  
  <meta name="twitter:description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  

<meta name="author" content="Stephen Woloszynek" />


<meta name="date" content="2020-02-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="qiime.html"/>
<link rel="next" href="lasso.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li>Data Science Examples</li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="dynprog.html"><a href="dynprog.html"><i class="fa fa-check"></i><b>2</b> Dynamic Programming</a><ul>
<li class="chapter" data-level="2.1" data-path="dynprog.html"><a href="dynprog.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="dynprog.html"><a href="dynprog.html#rod-cutting"><i class="fa fa-check"></i><b>2.2</b> Rod cutting</a></li>
<li class="chapter" data-level="2.3" data-path="dynprog.html"><a href="dynprog.html#fibonacci-rabbits"><i class="fa fa-check"></i><b>2.3</b> Fibonacci rabbits</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="align.html"><a href="align.html"><i class="fa fa-check"></i><b>3</b> Alignment</a><ul>
<li class="chapter" data-level="3.1" data-path="align.html"><a href="align.html#longest-common-subsequence"><i class="fa fa-check"></i><b>3.1</b> Longest Common Subsequence</a></li>
<li class="chapter" data-level="3.2" data-path="align.html"><a href="align.html#global-alignment"><i class="fa fa-check"></i><b>3.2</b> Global Alignment</a></li>
<li class="chapter" data-level="3.3" data-path="align.html"><a href="align.html#local-alignment"><i class="fa fa-check"></i><b>3.3</b> Local Alignment</a></li>
<li class="chapter" data-level="3.4" data-path="align.html"><a href="align.html#local-alignment-homework"><i class="fa fa-check"></i><b>3.4</b> Local Alignment: Homework</a></li>
<li class="chapter" data-level="3.5" data-path="align.html"><a href="align.html#global-alignment-code-r"><i class="fa fa-check"></i><b>3.5</b> Global Alignment Code (R)</a></li>
<li class="chapter" data-level="3.6" data-path="align.html"><a href="align.html#global-alignment-code-python"><i class="fa fa-check"></i><b>3.6</b> Global Alignment Code (Python)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="alignalg.html"><a href="alignalg.html"><i class="fa fa-check"></i><b>4</b> Alignment Algorithms</a><ul>
<li class="chapter" data-level="4.1" data-path="alignalg.html"><a href="alignalg.html#longest-common-subsequence-1"><i class="fa fa-check"></i><b>4.1</b> Longest Common Subsequence</a></li>
<li class="chapter" data-level="4.2" data-path="alignalg.html"><a href="alignalg.html#global-alignment-r"><i class="fa fa-check"></i><b>4.2</b> Global Alignment (R)</a></li>
<li class="chapter" data-level="4.3" data-path="alignalg.html"><a href="alignalg.html#global-alignment-python"><i class="fa fa-check"></i><b>4.3</b> Global Alignment (Python)</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="biocond.html"><a href="biocond.html"><i class="fa fa-check"></i><b>5</b> Bioconductor</a><ul>
<li class="chapter" data-level="5.0.1" data-path="biocond.html"><a href="biocond.html#loading-fasta-files"><i class="fa fa-check"></i><b>5.0.1</b> Loading FASTA Files</a></li>
<li class="chapter" data-level="5.0.2" data-path="biocond.html"><a href="biocond.html#creating-sequence-sets"><i class="fa fa-check"></i><b>5.0.2</b> Creating Sequence Sets</a></li>
<li class="chapter" data-level="5.0.3" data-path="biocond.html"><a href="biocond.html#sample-metadata"><i class="fa fa-check"></i><b>5.0.3</b> Sample Metadata</a></li>
<li class="chapter" data-level="5.1" data-path="biocond.html"><a href="biocond.html#creating-gc-functions"><i class="fa fa-check"></i><b>5.1</b> Creating GC Functions</a></li>
<li class="chapter" data-level="5.2" data-path="biocond.html"><a href="biocond.html#ncbi-esearch"><i class="fa fa-check"></i><b>5.2</b> NCBI ESearch</a></li>
<li class="chapter" data-level="5.3" data-path="biocond.html"><a href="biocond.html#cds"><i class="fa fa-check"></i><b>5.3</b> CDS</a></li>
<li class="chapter" data-level="5.4" data-path="biocond.html"><a href="biocond.html#whole-genomes"><i class="fa fa-check"></i><b>5.4</b> Whole Genomes</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>6</b> Retrieving Projects</a><ul>
<li class="chapter" data-level="6.0.1" data-path="sra.html"><a href="sra.html#fastq-dump-for-paired-end-reads"><i class="fa fa-check"></i><b>6.0.1</b> Fastq Dump for Paired End Reads</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="phyloseq.html"><a href="phyloseq.html"><i class="fa fa-check"></i><b>7</b> Phyloseq</a></li>
<li class="chapter" data-level="8" data-path="dada.html"><a href="dada.html"><i class="fa fa-check"></i><b>8</b> Dada2</a><ul>
<li class="chapter" data-level="8.1" data-path="dada.html"><a href="dada.html#fastq-prep"><i class="fa fa-check"></i><b>8.1</b> FASTQ Prep</a></li>
<li class="chapter" data-level="8.2" data-path="dada.html"><a href="dada.html#otu-picking"><i class="fa fa-check"></i><b>8.2</b> OTU Picking</a></li>
<li class="chapter" data-level="8.3" data-path="dada.html"><a href="dada.html#running-dada2-on-proteus"><i class="fa fa-check"></i><b>8.3</b> Running Dada2 on Proteus</a><ul>
<li class="chapter" data-level="8.3.1" data-path="dada.html"><a href="dada.html#method-1-using-namegrp-shared-r-library"><i class="fa fa-check"></i><b>8.3.1</b> Method 1: Using nameGrp Shared R Library</a></li>
<li class="chapter" data-level="8.3.2" data-path="dada.html"><a href="dada.html#method-2-creating-a-local-library"><i class="fa fa-check"></i><b>8.3.2</b> Method 2: Creating a Local Library</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="qiime.html"><a href="qiime.html"><i class="fa fa-check"></i><b>9</b> Qiime</a><ul>
<li class="chapter" data-level="9.1" data-path="qiime.html"><a href="qiime.html#otu-picking-1"><i class="fa fa-check"></i><b>9.1</b> OTU Picking</a></li>
<li class="chapter" data-level="9.2" data-path="qiime.html"><a href="qiime.html#summarizing-our-results"><i class="fa fa-check"></i><b>9.2</b> Summarizing Our Results</a></li>
<li class="chapter" data-level="9.3" data-path="qiime.html"><a href="qiime.html#loading-qiime-results-into-phyloseq"><i class="fa fa-check"></i><b>9.3</b> Loading QIIME Results into Phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="multcomp.html"><a href="multcomp.html"><i class="fa fa-check"></i><b>10</b> Multiple Comparisons</a><ul>
<li class="chapter" data-level="10.1" data-path="multcomp.html"><a href="multcomp.html#hypothesis-testing-and-power"><i class="fa fa-check"></i><b>10.1</b> Hypothesis Testing and Power</a></li>
<li class="chapter" data-level="10.2" data-path="multcomp.html"><a href="multcomp.html#multiple-comparisons"><i class="fa fa-check"></i><b>10.2</b> Multiple Comparisons</a><ul>
<li class="chapter" data-level="10.2.1" data-path="multcomp.html"><a href="multcomp.html#bonferroni"><i class="fa fa-check"></i><b>10.2.1</b> Bonferroni</a></li>
<li class="chapter" data-level="10.2.2" data-path="multcomp.html"><a href="multcomp.html#false-discovery-rate"><i class="fa fa-check"></i><b>10.2.2</b> False Discovery Rate</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="lasso.html"><a href="lasso.html"><i class="fa fa-check"></i><b>11</b> Lasso</a><ul>
<li class="chapter" data-level="11.1" data-path="lasso.html"><a href="lasso.html#big-data-and-feature-selection"><i class="fa fa-check"></i><b>11.1</b> Big Data and Feature Selection</a></li>
<li class="chapter" data-level="11.2" data-path="lasso.html"><a href="lasso.html#lasso-regression"><i class="fa fa-check"></i><b>11.2</b> Lasso Regression</a></li>
<li class="chapter" data-level="11.3" data-path="lasso.html"><a href="lasso.html#multivariate-lasso-regression"><i class="fa fa-check"></i><b>11.3</b> Multivariate Lasso Regression</a></li>
<li class="chapter" data-level="11.4" data-path="lasso.html"><a href="lasso.html#parallel-coordinate-descent"><i class="fa fa-check"></i><b>11.4</b> Parallel Coordinate Descent</a></li>
<li class="chapter" data-level="11.5" data-path="lasso.html"><a href="lasso.html#simulations"><i class="fa fa-check"></i><b>11.5</b> Simulations</a><ul>
<li class="chapter" data-level="11.5.1" data-path="lasso.html"><a href="lasso.html#times-1000-matrix-5-target-coefficients"><i class="fa fa-check"></i><b>11.5.1</b> <span class="math inline">\(50 \times 1000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.2" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-5-target-coefficients"><i class="fa fa-check"></i><b>11.5.2</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.3" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-5-target-coefficients-1"><i class="fa fa-check"></i><b>11.5.3</b> <span class="math inline">\(200 \times 10000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.4" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-15-target-coefficients"><i class="fa fa-check"></i><b>11.5.4</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 15 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.5" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-30-target-coefficients"><i class="fa fa-check"></i><b>11.5.5</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 30 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.6" data-path="lasso.html"><a href="lasso.html#scaling-the-coeficients"><i class="fa fa-check"></i><b>11.5.6</b> Scaling the coeficients</a></li>
<li class="chapter" data-level="11.5.7" data-path="lasso.html"><a href="lasso.html#comparison"><i class="fa fa-check"></i><b>11.5.7</b> Comparison</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="lasso.html"><a href="lasso.html#conclusion"><i class="fa fa-check"></i><b>11.6</b> Conclusion</a></li>
<li class="chapter" data-level="11.7" data-path="lasso.html"><a href="lasso.html#code-lasso"><i class="fa fa-check"></i><b>11.7</b> Code: Lasso</a></li>
<li class="chapter" data-level="11.8" data-path="lasso.html"><a href="lasso.html#code-lasso-via-cyclic-gradient-descent"><i class="fa fa-check"></i><b>11.8</b> Code: Lasso via cyclic gradient descent</a></li>
<li class="chapter" data-level="11.9" data-path="lasso.html"><a href="lasso.html#code-parallel-lasso"><i class="fa fa-check"></i><b>11.9</b> Code: Parallel lasso</a></li>
<li class="chapter" data-level="11.10" data-path="lasso.html"><a href="lasso.html#references"><i class="fa fa-check"></i><b>11.10</b> References</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="tm.html"><a href="tm.html"><i class="fa fa-check"></i><b>12</b> Topic Models</a><ul>
<li class="chapter" data-level="12.1" data-path="tm.html"><a href="tm.html#variational-inference"><i class="fa fa-check"></i><b>12.1</b> Variational Inference</a><ul>
<li class="chapter" data-level="12.1.1" data-path="tm.html"><a href="tm.html#evidence-lower-bound-elbo"><i class="fa fa-check"></i><b>12.1.1</b> Evidence Lower Bound (ELBO)</a></li>
<li class="chapter" data-level="12.1.2" data-path="tm.html"><a href="tm.html#elbo-and-kl-divergence"><i class="fa fa-check"></i><b>12.1.2</b> ELBO and KL Divergence</a></li>
<li class="chapter" data-level="12.1.3" data-path="tm.html"><a href="tm.html#mean-field-method"><i class="fa fa-check"></i><b>12.1.3</b> Mean Field Method</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="tm.html"><a href="tm.html#lda"><i class="fa fa-check"></i><b>12.2</b> LDA</a><ul>
<li class="chapter" data-level="12.2.1" data-path="tm.html"><a href="tm.html#expectation-of-pthetaalpha"><i class="fa fa-check"></i><b>12.2.1</b> Expectation of p(<span class="math inline">\(\theta|\alpha)\)</span></a></li>
<li class="chapter" data-level="12.2.2" data-path="tm.html"><a href="tm.html#expectation-of-pztheta"><i class="fa fa-check"></i><b>12.2.2</b> Expectation of p(<span class="math inline">\(z|\theta)\)</span></a></li>
<li class="chapter" data-level="12.2.3" data-path="tm.html"><a href="tm.html#expectation-of-pwzbeta"><i class="fa fa-check"></i><b>12.2.3</b> Expectation of p(<span class="math inline">\(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.2.4" data-path="tm.html"><a href="tm.html#entropy-of-gamma-and-phi"><i class="fa fa-check"></i><b>12.2.4</b> Entropy of <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.2.5" data-path="tm.html"><a href="tm.html#complete-objective-function"><i class="fa fa-check"></i><b>12.2.5</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.2.6" data-path="tm.html"><a href="tm.html#parameter-optimization"><i class="fa fa-check"></i><b>12.2.6</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="tm.html"><a href="tm.html#supervised-lda"><i class="fa fa-check"></i><b>12.3</b> Supervised LDA</a><ul>
<li class="chapter" data-level="12.3.1" data-path="tm.html"><a href="tm.html#expectations-of-pthetaalpha-pztheta-and-pwzbeta"><i class="fa fa-check"></i><b>12.3.1</b> Expectations of <span class="math inline">\(p(\theta|\alpha)\)</span>, <span class="math inline">\(p(z|\theta)\)</span>, and <span class="math inline">\(p(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.3.2" data-path="tm.html"><a href="tm.html#expectation-of-pyzetasigma2"><i class="fa fa-check"></i><b>12.3.2</b> Expectation of <span class="math inline">\(p(y|z,\eta,\sigma^2)\)</span></a></li>
<li class="chapter" data-level="12.3.3" data-path="tm.html"><a href="tm.html#entropy-of-gamma-and-phi-1"><i class="fa fa-check"></i><b>12.3.3</b> Entropy of <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.3.4" data-path="tm.html"><a href="tm.html#complete-objective-function-1"><i class="fa fa-check"></i><b>12.3.4</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.3.5" data-path="tm.html"><a href="tm.html#parameter-optimization-1"><i class="fa fa-check"></i><b>12.3.5</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="tm.html"><a href="tm.html#the-correlated-topic-model"><i class="fa fa-check"></i><b>12.4</b> The Correlated Topic Model</a><ul>
<li class="chapter" data-level="12.4.1" data-path="tm.html"><a href="tm.html#multinomial-distribution-in-exponential-form"><i class="fa fa-check"></i><b>12.4.1</b> Multinomial Distribution in Exponential Form</a></li>
<li class="chapter" data-level="12.4.2" data-path="tm.html"><a href="tm.html#variational-em"><i class="fa fa-check"></i><b>12.4.2</b> Variational EM</a></li>
<li class="chapter" data-level="12.4.3" data-path="tm.html"><a href="tm.html#expectation-of-pwzbeta-1"><i class="fa fa-check"></i><b>12.4.3</b> Expectation of <span class="math inline">\(p(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.4.4" data-path="tm.html"><a href="tm.html#expectation-of-pzeta"><i class="fa fa-check"></i><b>12.4.4</b> Expectation of <span class="math inline">\(p(z|\eta)\)</span></a></li>
<li class="chapter" data-level="12.4.5" data-path="tm.html"><a href="tm.html#expectation-of-petamusigma"><i class="fa fa-check"></i><b>12.4.5</b> Expectation of <span class="math inline">\(p(\eta|\mu,\sigma)\)</span></a></li>
<li class="chapter" data-level="12.4.6" data-path="tm.html"><a href="tm.html#entropy-of-lambda-nu-and-phi"><i class="fa fa-check"></i><b>12.4.6</b> Entropy of <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\nu\)</span>, and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.4.7" data-path="tm.html"><a href="tm.html#complete-objective-function-2"><i class="fa fa-check"></i><b>12.4.7</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.4.8" data-path="tm.html"><a href="tm.html#parameter-optimization-2"><i class="fa fa-check"></i><b>12.4.8</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="tm.html"><a href="tm.html#dirichlet-distribution"><i class="fa fa-check"></i><b>12.5</b> Dirichlet Distribution</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="ml.html"><a href="ml.html"><i class="fa fa-check"></i><b>13</b> Machine Learning</a><ul>
<li class="chapter" data-level="13.1" data-path="ml.html"><a href="ml.html#cross-valdiation"><i class="fa fa-check"></i><b>13.1</b> Cross Valdiation</a><ul>
<li class="chapter" data-level="13.1.1" data-path="ml.html"><a href="ml.html#loocv"><i class="fa fa-check"></i><b>13.1.1</b> LOOCV</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="ml.html"><a href="ml.html#naive-bayes"><i class="fa fa-check"></i><b>13.2</b> Naive Bayes</a></li>
<li class="chapter" data-level="13.3" data-path="ml.html"><a href="ml.html#svm"><i class="fa fa-check"></i><b>13.3</b> SVM</a><ul>
<li class="chapter" data-level="13.3.1" data-path="ml.html"><a href="ml.html#manual-rbf-kernal"><i class="fa fa-check"></i><b>13.3.1</b> Manual rbf kernal</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="ml.html"><a href="ml.html#k-means"><i class="fa fa-check"></i><b>13.4</b> K-means</a></li>
<li class="chapter" data-level="13.5" data-path="ml.html"><a href="ml.html#gaussian-mixtures"><i class="fa fa-check"></i><b>13.5</b> Gaussian Mixtures</a></li>
<li class="chapter" data-level="13.6" data-path="ml.html"><a href="ml.html#pca"><i class="fa fa-check"></i><b>13.6</b> PCA</a></li>
<li class="chapter" data-level="13.7" data-path="ml.html"><a href="ml.html#viterbi-algorithm"><i class="fa fa-check"></i><b>13.7</b> Viterbi Algorithm</a></li>
<li class="chapter" data-level="13.8" data-path="ml.html"><a href="ml.html#gradient-descent"><i class="fa fa-check"></i><b>13.8</b> Gradient Descent</a><ul>
<li class="chapter" data-level="13.8.1" data-path="ml.html"><a href="ml.html#linear-regression"><i class="fa fa-check"></i><b>13.8.1</b> Linear Regression</a></li>
<li class="chapter" data-level="13.8.2" data-path="ml.html"><a href="ml.html#logistic-regression"><i class="fa fa-check"></i><b>13.8.2</b> Logistic Regression</a></li>
<li class="chapter" data-level="13.8.3" data-path="ml.html"><a href="ml.html#softmax-regression"><i class="fa fa-check"></i><b>13.8.3</b> Softmax regression</a></li>
</ul></li>
<li class="chapter" data-level="13.9" data-path="ml.html"><a href="ml.html#nonparametric-bayesian-processes"><i class="fa fa-check"></i><b>13.9</b> Nonparametric Bayesian Processes</a><ul>
<li class="chapter" data-level="13.9.1" data-path="ml.html"><a href="ml.html#chinese-restaurant"><i class="fa fa-check"></i><b>13.9.1</b> Chinese Restaurant</a></li>
<li class="chapter" data-level="13.9.2" data-path="ml.html"><a href="ml.html#polyas-urn"><i class="fa fa-check"></i><b>13.9.2</b> Polyas Urn</a></li>
<li class="chapter" data-level="13.9.3" data-path="ml.html"><a href="ml.html#stick-breaking"><i class="fa fa-check"></i><b>13.9.3</b> Stick Breaking</a></li>
</ul></li>
<li class="chapter" data-level="13.10" data-path="ml.html"><a href="ml.html#iteratively-reweighted-least-squares"><i class="fa fa-check"></i><b>13.10</b> Iteratively Reweighted Least Squares</a></li>
<li class="chapter" data-level="13.11" data-path="ml.html"><a href="ml.html#neural-network"><i class="fa fa-check"></i><b>13.11</b> Neural Network</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/sw1/" target="blank">Published by Stephen Woloszynek</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Worked Bioninformatics, Statistics, and Machine Learning Examples</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="multcomp" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Multiple Comparisons</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(shiny)</code></pre></div>
<div id="hypothesis-testing-and-power" class="section level2">
<h2><span class="header-section-number">10.1</span> Hypothesis Testing and Power</h2>
<p>When we perform a statistical hypothesis test, we measure a given effect, assess its probability of occuring given a null hypothesis <span class="math inline">\(H_0\)</span> (i.e., the hypothesis that there is <em>no</em> relationship between our test groupings), and reject said null if the probability falls below a particular threshold. We “accept” <span class="math inline">\(H_1\)</span> and consider this test to be <em>statistically significant</em>. We aren’t always right, however; hence, hypothesis testing is prone to <em>false positives (type I errors) and false negatives (type II errors)</em>:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left"><strong>Reject</strong> <span class="math inline">\(H_0\)</span></th>
<th align="left"><strong>Reject</strong> <span class="math inline">\(H_1\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><strong>False</strong> <span class="math inline">\(H_0\)</span></td>
<td align="left">True Positives (TP)</td>
<td align="left">False Negatives (FN)</td>
</tr>
<tr class="even">
<td align="left"><strong>True</strong> <span class="math inline">\(H_0\)</span></td>
<td align="left">False Positives (FP)</td>
<td align="left">True Negatives (TN)</td>
</tr>
</tbody>
</table>
<p>You’ll often see the false positive error probability refered to as <span class="math inline">\(\alpha\)</span> (where <span class="math inline">\(\alpha=\frac{\text{FP}}{\text{FP}+\text{TN}}\)</span>), the false negative error probability referred to as <span class="math inline">\(\beta\)</span> (where <span class="math inline">\(\beta=\frac{\text{FN}}{\text{FN}+\text{TP}}\)</span>), and the <strong>probabilty of a false negative not occuring</strong> as <span class="math inline">\(1-\beta\)</span>. Tests that have low <span class="math inline">\(beta\)</span> are said to have high <em>power</em>, where <span class="math inline">\(\text{Power}=1-\beta\)</span>. One can interpret a powerful test as a test that is very capable of detecting when the null hypothesis is false – that is, <span class="math inline">\(P(\text{reject } H_0|\text{true } H_1)\)</span>. If we were comparing two groups, and our power was about 0.9, then we’d have about a 90% chance of detecting a difference between the groups and a 10% chance of missing it.</p>
<iframe src="https://sw424.shinyapps.io/power/?showcase=0" width="768" height="700">
</iframe>
</div>
<div id="multiple-comparisons" class="section level2">
<h2><span class="header-section-number">10.2</span> Multiple Comparisons</h2>
<p>Whether you continue to work with large datasets or transition into the types of data problems one often sees stemming from benchwork, you’ll come across the multiple comparisons problem. Let’s focus on a situation common when working with DNA or RNA sequencing counts. Say we were given some <span class="math inline">\(N \times M\)</span> normalized gene table, where there are <span class="math inline">\(N=100\)</span> samples and <span class="math inline">\(M=10000\)</span> genes. Of these samples, 50 will belong to group 1 and the other 50 to group 2:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">N_grp1 &lt;-<span class="st"> </span><span class="dv">50</span>
N_grp2 &lt;-<span class="st"> </span><span class="dv">50</span>
N &lt;-<span class="st"> </span>N_grp1+N_grp2
M &lt;-<span class="st"> </span><span class="dv">10000</span>

<span class="kw">set.seed</span>(<span class="dv">453</span>)
TABLE_grp1 &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(N_grp1*M,<span class="dv">0</span>,<span class="dv">1</span>),N_grp1,M,<span class="dt">dimnames=</span><span class="kw">list</span>(<span class="kw">paste0</span>(<span class="st">&#39;sample&#39;</span>,<span class="dv">1</span>:N_grp1),<span class="kw">paste0</span>(<span class="st">&#39;gene&#39;</span>,<span class="dv">1</span>:M)))
TABLE_grp2 &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(N_grp2*M,<span class="dv">0</span>,<span class="dv">1</span>),N_grp2,M,<span class="dt">dimnames=</span><span class="kw">list</span>(<span class="kw">paste0</span>(<span class="st">&#39;sample&#39;</span>,<span class="dv">1</span>:N_grp2),<span class="kw">paste0</span>(<span class="st">&#39;gene&#39;</span>,<span class="dv">1</span>:M)))
TABLE &lt;-<span class="st"> </span><span class="kw">rbind</span>(TABLE_grp1,TABLE_grp2)</code></pre></div>
<p>Now, let’s aim to identify important genes that are different between the two groups. We can do this by performing a t test between the groups for each gene sequentially. I’m going to hardcode the t test there in case you are unfamiliar.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tstats &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="dt">mode=</span><span class="st">&#39;double&#39;</span>,<span class="dt">length=</span>M)
for (gene in <span class="kw">seq_len</span>(M)){
  grp1 &lt;-<span class="st"> </span>TABLE[<span class="dv">1</span>:<span class="dv">50</span>,gene]
  grp2 &lt;-<span class="st"> </span>TABLE[<span class="dv">51</span>:<span class="dv">100</span>,gene]
  
  mu1 &lt;-<span class="st"> </span><span class="kw">mean</span>(grp1)
  mu2 &lt;-<span class="st"> </span><span class="kw">mean</span>(grp2)
  
  v1 &lt;-<span class="st"> </span><span class="kw">sum</span>((grp1 -<span class="st"> </span><span class="kw">mean</span>(grp1))^<span class="dv">2</span>)
  v2 &lt;-<span class="st"> </span><span class="kw">sum</span>((grp2 -<span class="st"> </span><span class="kw">mean</span>(grp2))^<span class="dv">2</span>)
  
  s2 &lt;-<span class="st"> </span>(<span class="dv">1</span>/(N<span class="dv">-2</span>))*(v1 +<span class="st"> </span>v2)
  se &lt;-<span class="st"> </span><span class="kw">sqrt</span>(s2*(<span class="dv">1</span>/N_grp1 +<span class="st"> </span><span class="dv">1</span>/N_grp2))
  
  tstats[gene] &lt;-<span class="st"> </span>(mu1-mu2)/se
}</code></pre></div>
<p>And now we can plot it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(tstats,<span class="dt">geom=</span><span class="st">&#39;histogram&#39;</span>,<span class="dt">color=</span><span class="kw">I</span>(<span class="st">&#39;black&#39;</span>),<span class="dt">bins=</span><span class="dv">50</span>)</code></pre></div>
<p><img src="09_multiple_comparisons_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>Clearly the t statistics are normally distributed. The typical approach with t tests is to perform a hypothesis test at the 5% significance level, such that we reject any null hypothesis with t statistics more than 2 standard deviations from the mode in either direction of this distribution. We use 2 standard deviations because qnorm(0.975) is 1.96 and qnorm(0.025) is -1.96:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(tstats,<span class="dt">geom=</span><span class="st">&#39;histogram&#39;</span>,<span class="dt">color=</span><span class="kw">I</span>(<span class="st">&#39;black&#39;</span>),<span class="dt">bins=</span><span class="dv">50</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span><span class="kw">qnorm</span>(<span class="fl">0.025</span>),<span class="dt">linetype=</span><span class="dv">2</span>,<span class="dt">color=</span><span class="st">&#39;red&#39;</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span><span class="kw">qnorm</span>(<span class="fl">0.975</span>),<span class="dt">linetype=</span><span class="dv">2</span>,<span class="dt">color=</span><span class="st">&#39;red&#39;</span>)</code></pre></div>
<p><img src="09_multiple_comparisons_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>It should be apparent that there are quite a few significant genes. If you were paying attention to how our groups were generated, this should be surprising to you. We generated both groups from the <em>same</em> normal distribution. Nevertheless, we still have a ton of significant genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(tstats&lt;<span class="kw">qnorm</span>(<span class="fl">0.025</span>) |<span class="st"> </span>tstats&gt;<span class="kw">qnorm</span>(<span class="fl">0.975</span>))</code></pre></div>
<pre><code>## [1] 530</code></pre>
<p>So over 500 genes were significant despite the fact that we know there is no underlying effect. Well, this is actually quite consistent with the defiinition of a p value. Given a typical hypothesis test, we reject the null if the probability of the observation is less than the probability of the observation occuring assuming the null hypothesis is true. In other words, we reject the claim that there is no statistical effect if our p value is less than (in absolute value) <span class="math inline">\(\alpha\)</span>, the probability of a statistical effect due to random sampling variation. Therefore, because there are 10,000 genes and hence we perform 10,000 t tests, givin an <span class="math inline">\(\alpha\)</span> of 0.05, we should expect about <span class="math inline">\(0.05 \times 10000 = 500\)</span> significant t statistics due to random sampling variation alone, which is the case here.</p>
<p>So that’s all well and good. We now know that if we set <span class="math inline">\(\alpha\)</span> at a particular level, given our null and sample size, we should expect some false positives. But is that enough information? Let’s rerun the above model, but now with an actual difference between groups:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">N_grp1 &lt;-<span class="st"> </span><span class="dv">50</span>
N_grp2 &lt;-<span class="st"> </span><span class="dv">50</span>
N &lt;-<span class="st"> </span>N_grp1+N_grp2
M &lt;-<span class="st"> </span><span class="dv">10000</span>

<span class="kw">set.seed</span>(<span class="dv">65</span>)
TABLE_grp1 &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(N_grp1*M,.<span class="dv">25</span>,<span class="dv">2</span>),N_grp1,M,<span class="dt">dimnames=</span><span class="kw">list</span>(<span class="kw">paste0</span>(<span class="st">&#39;sample&#39;</span>,<span class="dv">1</span>:N_grp1),<span class="kw">paste0</span>(<span class="st">&#39;gene&#39;</span>,<span class="dv">1</span>:M)))
TABLE_grp2 &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(N_grp2*M,-.<span class="dv">25</span>,<span class="dv">2</span>),N_grp2,M,<span class="dt">dimnames=</span><span class="kw">list</span>(<span class="kw">paste0</span>(<span class="st">&#39;sample&#39;</span>,<span class="dv">1</span>:N_grp2),<span class="kw">paste0</span>(<span class="st">&#39;gene&#39;</span>,<span class="dv">1</span>:M)))
TABLE &lt;-<span class="st"> </span><span class="kw">rbind</span>(TABLE_grp1,TABLE_grp2)

tstats &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="dt">mode=</span><span class="st">&#39;double&#39;</span>,<span class="dt">length=</span>M)
tstats2 &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="dt">mode=</span><span class="st">&#39;double&#39;</span>,<span class="dt">length=</span>M)
for (gene in <span class="kw">seq_len</span>(M)){
  grp1 &lt;-<span class="st"> </span>TABLE[<span class="dv">1</span>:<span class="dv">50</span>,gene]
  grp2 &lt;-<span class="st"> </span>TABLE[<span class="dv">51</span>:<span class="dv">100</span>,gene]
  
  mu1 &lt;-<span class="st"> </span><span class="kw">mean</span>(grp1)
  mu2 &lt;-<span class="st"> </span><span class="kw">mean</span>(grp2)
  
  v1 &lt;-<span class="st"> </span><span class="kw">sum</span>((grp1 -<span class="st"> </span><span class="kw">mean</span>(grp1))^<span class="dv">2</span>)
  v2 &lt;-<span class="st"> </span><span class="kw">sum</span>((grp2 -<span class="st"> </span><span class="kw">mean</span>(grp2))^<span class="dv">2</span>)
  
  s2 &lt;-<span class="st"> </span>(<span class="dv">1</span>/(N<span class="dv">-2</span>))*(v1 +<span class="st"> </span>v2)
  se &lt;-<span class="st"> </span><span class="kw">sqrt</span>(s2*(<span class="dv">1</span>/N_grp1 +<span class="st"> </span><span class="dv">1</span>/N_grp2))
  
  tstats[gene] &lt;-<span class="st"> </span>(mu1-mu2)/se
}

<span class="kw">sum</span>(tstats&lt;<span class="kw">qnorm</span>(<span class="fl">0.025</span>) |<span class="st"> </span>tstats&gt;<span class="kw">qnorm</span>(<span class="fl">0.975</span>))</code></pre></div>
<pre><code>## [1] 2386</code></pre>
<p>Now the two groups have different normlized gene values, which leads to over 2000 significant genes. We know from before that about 500 should be due to random sampling variation, but how do we know which ones? This brings us to methods to adjust our results such that we can interpret them more easily.</p>
<div id="bonferroni" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Bonferroni</h3>
<p>We’ll start with Bonferroni correction. First thing we’ll do is calculate the p values for all of those t statistics:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pvals &lt;-<span class="st"> </span><span class="dv">2</span>*(<span class="dv">1</span>-<span class="kw">pt</span>(tstats,N<span class="dv">-2</span>))</code></pre></div>
<p>Let’s continue to assume <span class="math inline">\(\alpha=0.05\)</span>. Now, we should have <em>about</em> the same number of significant genes as mentioned before:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha &lt;-<span class="st"> </span>.<span class="dv">05</span>
<span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha)</code></pre></div>
<pre><code>## [1] 2309</code></pre>
<p>Bonferroni corrects for multiple comparisons by controlling for the <em>family-wise error rate (FWER)</em> – the probability of making at least one false positive – by simply dividing <span class="math inline">\(\alpha\)</span> by the number of tests performed, which gives a new, adjusted <span class="math inline">\(\alpha\)</span> to use as a significance threshold:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha_bonf &lt;-<span class="st"> </span>alpha/M
<span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha_bonf)</code></pre></div>
<pre><code>## [1] 5</code></pre>
<p>This gives us only 2 significant genes. We should expect more given the way we set the group means. The issue with using Bonferonni here is that it’s quite <strong>conservative</strong>, and tends to really <strong>decrease statistical power</strong> when there are a lot of tests, as in this case. Consequently, we are likely to see a ton of <strong>false negatives</strong>. In the cases where we perform only a few tests (say 10 or less), Bonferonni is quick and easy, and probably appropriate, but definitely not here.</p>
</div>
<div id="false-discovery-rate" class="section level3">
<h3><span class="header-section-number">10.2.2</span> False Discovery Rate</h3>
<p>Instead of controlling for the family-wise error rate, we can control for the <em>false discovery rate (FDR)</em>, the expected proportion of false positives. Unlike the FWER, which provides stringent control over false positives, the FDR is far more lenient, thereby <em>increasing power at the sacrifice for more false positives</em>.</p>
<p>If we rewrite our table from above as follows:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left"><strong>Reject</strong> <span class="math inline">\(H_0\)</span></th>
<th align="left"><strong>Reject</strong> <span class="math inline">\(H_1\)</span></th>
<th align="left"><strong>Total</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><strong>False</strong> <span class="math inline">\(H_0\)</span></td>
<td align="left">True Positives (TP)</td>
<td align="left">False Negatives (FN)</td>
<td align="left">Actual Positives (AP)</td>
</tr>
<tr class="even">
<td align="left"><strong>True</strong> <span class="math inline">\(H_0\)</span></td>
<td align="left">False Positives (FP)</td>
<td align="left">True Negatives (TN)</td>
<td align="left">Actual Negatives (AN)</td>
</tr>
<tr class="odd">
<td align="left"><strong>Total</strong></td>
<td align="left">Called Positives (CP)</td>
<td align="left">Called Negatives (CN)</td>
<td align="left">Total (TOTAL)</td>
</tr>
</tbody>
</table>
<p>we can rewrite the FDR as <span class="math display">\[
\text{FDR}=\mathop{\mathbb{E}}\left[\frac{\text{FP}}{\text{CP}}\right]
\]</span>.</p>
<p>Assuming all of our tests are independent, and given a significance threshold <span class="math inline">\(\alpha\)</span>, the FDR has the following property</p>
<p><span class="math display">\[
\text{FDR}=\mathop{\mathbb{E}}\left[\frac{\text{FP}}{\text{CP}}\right] \le \alpha\frac{\text{AN}}{\text{TOTAL}} \le \alpha
\]</span> Below is a figure to show what FDR correction will ultimately do to our dataset we generated above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha &lt;-<span class="st"> </span>.<span class="dv">05</span>
df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gene=</span><span class="dv">1</span>:<span class="kw">length</span>(pvals),
                 <span class="dt">p=</span><span class="kw">sort</span>(pvals),
                 <span class="dt">fit=</span>(<span class="dv">1</span>:<span class="kw">length</span>(pvals))*alpha/M)
<span class="kw">ggplot</span>(df,<span class="kw">aes</span>(gene,p)) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span><span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha),<span class="dt">size=</span>.<span class="dv">5</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span>alpha*M,<span class="dt">size=</span>.<span class="dv">5</span>) +
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(gene,fit),<span class="dt">color=</span><span class="st">&#39;red&#39;</span>,<span class="dt">linetype=</span><span class="dv">3</span>,<span class="dt">size=</span>.<span class="dv">5</span>) +
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span>.<span class="dv">3</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">scale_x_log10</span>() +<span class="st"> </span><span class="kw">scale_y_log10</span>() +<span class="st"> </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="st">&#39;Genes&#39;</span>,<span class="dt">y=</span><span class="st">&#39;Log10 p Value&#39;</span>) +
<span class="st">  </span><span class="kw">annotate</span>(<span class="st">&#39;label&#39;</span>,<span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha),<span class="fl">1e-06</span>,<span class="dt">label=</span><span class="st">&#39;alpha&#39;</span>,<span class="dt">parse=</span><span class="ot">TRUE</span>) +
<span class="st">  </span><span class="kw">annotate</span>(<span class="st">&#39;label&#39;</span>,alpha*M,<span class="fl">1e-06</span>,<span class="dt">label=</span><span class="st">&#39;alpha*M&#39;</span>,<span class="dt">parse=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="09_multiple_comparisons_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>The genes are ordered in terms of increasing p value, with everything on log scale. The vertical lines shows the number of genes with <span class="math inline">\(p&lt;\alpha\)</span> and the number of genes we should expect based on random sampling alone, <span class="math inline">\(\alpha*M\)</span>, where M is our total number of genes. The red diagonal line is the gene index multiplied by <span class="math inline">\(\alpha/\text{TOTAL}\)</span>. This gives us the FDR threshold. Any genes that fall <em>below</em> this line are considered significant after FDR adjustment.</p>
<p>To perform this adjustment, we simply choose an alpha, then rank our M genes in terms of increasing p value and identify the largest p value for genes where</p>
<p><span class="math display">\[
p_\text{gene} &lt; \alpha \frac{\text{rank}}{M}
\]</span> This becomes our new significance threshold:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha_fdr &lt;-<span class="st"> </span><span class="kw">max</span>(df$p[df$p &lt;<span class="st"> </span>alpha*df$gene/M])
<span class="kw">sum</span>(pvals&lt;alpha_fdr)</code></pre></div>
<pre><code>## [1] 222</code></pre>
<p>This results in far more significant genes than the Bonferroni methods, but still less than half of what we’d expect from random sampling variation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(df,<span class="kw">aes</span>(gene,p)) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span><span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha),<span class="dt">size=</span>.<span class="dv">5</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span>alpha*M,<span class="dt">size=</span>.<span class="dv">5</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span><span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha_bonf),<span class="dt">size=</span>.<span class="dv">5</span>,<span class="dt">color=</span><span class="st">&#39;blue&#39;</span>) +
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept=</span><span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha_fdr),<span class="dt">size=</span>.<span class="dv">5</span>,<span class="dt">color=</span><span class="st">&#39;darkgreen&#39;</span>) +
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(gene,fit),<span class="dt">color=</span><span class="st">&#39;red&#39;</span>,<span class="dt">linetype=</span><span class="dv">3</span>,<span class="dt">size=</span>.<span class="dv">5</span>) +
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span>.<span class="dv">3</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">scale_x_log10</span>() +<span class="st"> </span><span class="kw">scale_y_log10</span>() +<span class="st"> </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="st">&#39;Genes&#39;</span>,<span class="dt">y=</span><span class="st">&#39;Log10 p Value&#39;</span>) +
<span class="st">  </span><span class="kw">annotate</span>(<span class="st">&#39;label&#39;</span>,<span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha),<span class="fl">1e-06</span>,<span class="dt">label=</span><span class="st">&#39;alpha&#39;</span>,<span class="dt">parse=</span><span class="ot">TRUE</span>) +
<span class="st">  </span><span class="kw">annotate</span>(<span class="st">&#39;label&#39;</span>,alpha*M,<span class="fl">1e-06</span>,<span class="dt">label=</span><span class="st">&#39;alpha*M&#39;</span>,<span class="dt">parse=</span><span class="ot">TRUE</span>) +
<span class="st">  </span><span class="kw">annotate</span>(<span class="st">&#39;label&#39;</span>,<span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha_bonf),<span class="fl">1e-06</span>,<span class="dt">label=</span><span class="st">&#39;alpha[bonf]&#39;</span>,<span class="dt">parse=</span><span class="ot">TRUE</span>) +
<span class="st">  </span><span class="kw">annotate</span>(<span class="st">&#39;label&#39;</span>,<span class="kw">sum</span>(pvals &lt;<span class="st"> </span>alpha_fdr),<span class="fl">1e-06</span>,<span class="dt">label=</span><span class="st">&#39;alpha[fdr]&#39;</span>,<span class="dt">parse=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="09_multiple_comparisons_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="qiime.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="lasso.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bioinformatics.pdf"],
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
