<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 Qiime | Worked Bioninformatics, Statistics, and Machine Learning Examples</title>
  <meta name="description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  <meta name="generator" content="bookdown 0.17 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 Qiime | Worked Bioninformatics, Statistics, and Machine Learning Examples" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://sw1.github.io/teaching/" />
  
  <meta property="og:description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  <meta name="github-repo" content="sw1/teaching" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 Qiime | Worked Bioninformatics, Statistics, and Machine Learning Examples" />
  
  <meta name="twitter:description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  

<meta name="author" content="Stephen Woloszynek" />


<meta name="date" content="2020-02-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="dada.html"/>
<link rel="next" href="multcomp.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li>Data Science Examples</li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="dynprog.html"><a href="dynprog.html"><i class="fa fa-check"></i><b>2</b> Dynamic Programming</a><ul>
<li class="chapter" data-level="2.1" data-path="dynprog.html"><a href="dynprog.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="dynprog.html"><a href="dynprog.html#rod-cutting"><i class="fa fa-check"></i><b>2.2</b> Rod cutting</a></li>
<li class="chapter" data-level="2.3" data-path="dynprog.html"><a href="dynprog.html#fibonacci-rabbits"><i class="fa fa-check"></i><b>2.3</b> Fibonacci rabbits</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="align.html"><a href="align.html"><i class="fa fa-check"></i><b>3</b> Alignment</a><ul>
<li class="chapter" data-level="3.1" data-path="align.html"><a href="align.html#longest-common-subsequence"><i class="fa fa-check"></i><b>3.1</b> Longest Common Subsequence</a></li>
<li class="chapter" data-level="3.2" data-path="align.html"><a href="align.html#global-alignment"><i class="fa fa-check"></i><b>3.2</b> Global Alignment</a></li>
<li class="chapter" data-level="3.3" data-path="align.html"><a href="align.html#local-alignment"><i class="fa fa-check"></i><b>3.3</b> Local Alignment</a></li>
<li class="chapter" data-level="3.4" data-path="align.html"><a href="align.html#local-alignment-homework"><i class="fa fa-check"></i><b>3.4</b> Local Alignment: Homework</a></li>
<li class="chapter" data-level="3.5" data-path="align.html"><a href="align.html#global-alignment-code-r"><i class="fa fa-check"></i><b>3.5</b> Global Alignment Code (R)</a></li>
<li class="chapter" data-level="3.6" data-path="align.html"><a href="align.html#global-alignment-code-python"><i class="fa fa-check"></i><b>3.6</b> Global Alignment Code (Python)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="alignalg.html"><a href="alignalg.html"><i class="fa fa-check"></i><b>4</b> Alignment Algorithms</a><ul>
<li class="chapter" data-level="4.1" data-path="alignalg.html"><a href="alignalg.html#longest-common-subsequence-1"><i class="fa fa-check"></i><b>4.1</b> Longest Common Subsequence</a></li>
<li class="chapter" data-level="4.2" data-path="alignalg.html"><a href="alignalg.html#global-alignment-r"><i class="fa fa-check"></i><b>4.2</b> Global Alignment (R)</a></li>
<li class="chapter" data-level="4.3" data-path="alignalg.html"><a href="alignalg.html#global-alignment-python"><i class="fa fa-check"></i><b>4.3</b> Global Alignment (Python)</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="biocond.html"><a href="biocond.html"><i class="fa fa-check"></i><b>5</b> Bioconductor</a><ul>
<li class="chapter" data-level="5.0.1" data-path="biocond.html"><a href="biocond.html#loading-fasta-files"><i class="fa fa-check"></i><b>5.0.1</b> Loading FASTA Files</a></li>
<li class="chapter" data-level="5.0.2" data-path="biocond.html"><a href="biocond.html#creating-sequence-sets"><i class="fa fa-check"></i><b>5.0.2</b> Creating Sequence Sets</a></li>
<li class="chapter" data-level="5.0.3" data-path="biocond.html"><a href="biocond.html#sample-metadata"><i class="fa fa-check"></i><b>5.0.3</b> Sample Metadata</a></li>
<li class="chapter" data-level="5.1" data-path="biocond.html"><a href="biocond.html#creating-gc-functions"><i class="fa fa-check"></i><b>5.1</b> Creating GC Functions</a></li>
<li class="chapter" data-level="5.2" data-path="biocond.html"><a href="biocond.html#ncbi-esearch"><i class="fa fa-check"></i><b>5.2</b> NCBI ESearch</a></li>
<li class="chapter" data-level="5.3" data-path="biocond.html"><a href="biocond.html#cds"><i class="fa fa-check"></i><b>5.3</b> CDS</a></li>
<li class="chapter" data-level="5.4" data-path="biocond.html"><a href="biocond.html#whole-genomes"><i class="fa fa-check"></i><b>5.4</b> Whole Genomes</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>6</b> Retrieving Projects</a><ul>
<li class="chapter" data-level="6.0.1" data-path="sra.html"><a href="sra.html#fastq-dump-for-paired-end-reads"><i class="fa fa-check"></i><b>6.0.1</b> Fastq Dump for Paired End Reads</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="phyloseq.html"><a href="phyloseq.html"><i class="fa fa-check"></i><b>7</b> Phyloseq</a></li>
<li class="chapter" data-level="8" data-path="dada.html"><a href="dada.html"><i class="fa fa-check"></i><b>8</b> Dada2</a><ul>
<li class="chapter" data-level="8.1" data-path="dada.html"><a href="dada.html#fastq-prep"><i class="fa fa-check"></i><b>8.1</b> FASTQ Prep</a></li>
<li class="chapter" data-level="8.2" data-path="dada.html"><a href="dada.html#otu-picking"><i class="fa fa-check"></i><b>8.2</b> OTU Picking</a></li>
<li class="chapter" data-level="8.3" data-path="dada.html"><a href="dada.html#running-dada2-on-proteus"><i class="fa fa-check"></i><b>8.3</b> Running Dada2 on Proteus</a><ul>
<li class="chapter" data-level="8.3.1" data-path="dada.html"><a href="dada.html#method-1-using-namegrp-shared-r-library"><i class="fa fa-check"></i><b>8.3.1</b> Method 1: Using nameGrp Shared R Library</a></li>
<li class="chapter" data-level="8.3.2" data-path="dada.html"><a href="dada.html#method-2-creating-a-local-library"><i class="fa fa-check"></i><b>8.3.2</b> Method 2: Creating a Local Library</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="qiime.html"><a href="qiime.html"><i class="fa fa-check"></i><b>9</b> Qiime</a><ul>
<li class="chapter" data-level="9.1" data-path="qiime.html"><a href="qiime.html#otu-picking-1"><i class="fa fa-check"></i><b>9.1</b> OTU Picking</a></li>
<li class="chapter" data-level="9.2" data-path="qiime.html"><a href="qiime.html#summarizing-our-results"><i class="fa fa-check"></i><b>9.2</b> Summarizing Our Results</a></li>
<li class="chapter" data-level="9.3" data-path="qiime.html"><a href="qiime.html#loading-qiime-results-into-phyloseq"><i class="fa fa-check"></i><b>9.3</b> Loading QIIME Results into Phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="multcomp.html"><a href="multcomp.html"><i class="fa fa-check"></i><b>10</b> Multiple Comparisons</a><ul>
<li class="chapter" data-level="10.1" data-path="multcomp.html"><a href="multcomp.html#hypothesis-testing-and-power"><i class="fa fa-check"></i><b>10.1</b> Hypothesis Testing and Power</a></li>
<li class="chapter" data-level="10.2" data-path="multcomp.html"><a href="multcomp.html#multiple-comparisons"><i class="fa fa-check"></i><b>10.2</b> Multiple Comparisons</a><ul>
<li class="chapter" data-level="10.2.1" data-path="multcomp.html"><a href="multcomp.html#bonferroni"><i class="fa fa-check"></i><b>10.2.1</b> Bonferroni</a></li>
<li class="chapter" data-level="10.2.2" data-path="multcomp.html"><a href="multcomp.html#false-discovery-rate"><i class="fa fa-check"></i><b>10.2.2</b> False Discovery Rate</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="lasso.html"><a href="lasso.html"><i class="fa fa-check"></i><b>11</b> Lasso</a><ul>
<li class="chapter" data-level="11.1" data-path="lasso.html"><a href="lasso.html#big-data-and-feature-selection"><i class="fa fa-check"></i><b>11.1</b> Big Data and Feature Selection</a></li>
<li class="chapter" data-level="11.2" data-path="lasso.html"><a href="lasso.html#lasso-regression"><i class="fa fa-check"></i><b>11.2</b> Lasso Regression</a></li>
<li class="chapter" data-level="11.3" data-path="lasso.html"><a href="lasso.html#multivariate-lasso-regression"><i class="fa fa-check"></i><b>11.3</b> Multivariate Lasso Regression</a></li>
<li class="chapter" data-level="11.4" data-path="lasso.html"><a href="lasso.html#parallel-coordinate-descent"><i class="fa fa-check"></i><b>11.4</b> Parallel Coordinate Descent</a></li>
<li class="chapter" data-level="11.5" data-path="lasso.html"><a href="lasso.html#simulations"><i class="fa fa-check"></i><b>11.5</b> Simulations</a><ul>
<li class="chapter" data-level="11.5.1" data-path="lasso.html"><a href="lasso.html#times-1000-matrix-5-target-coefficients"><i class="fa fa-check"></i><b>11.5.1</b> <span class="math inline">\(50 \times 1000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.2" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-5-target-coefficients"><i class="fa fa-check"></i><b>11.5.2</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.3" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-5-target-coefficients-1"><i class="fa fa-check"></i><b>11.5.3</b> <span class="math inline">\(200 \times 10000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.4" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-15-target-coefficients"><i class="fa fa-check"></i><b>11.5.4</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 15 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.5" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-30-target-coefficients"><i class="fa fa-check"></i><b>11.5.5</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 30 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.6" data-path="lasso.html"><a href="lasso.html#scaling-the-coeficients"><i class="fa fa-check"></i><b>11.5.6</b> Scaling the coeficients</a></li>
<li class="chapter" data-level="11.5.7" data-path="lasso.html"><a href="lasso.html#comparison"><i class="fa fa-check"></i><b>11.5.7</b> Comparison</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="lasso.html"><a href="lasso.html#conclusion"><i class="fa fa-check"></i><b>11.6</b> Conclusion</a></li>
<li class="chapter" data-level="11.7" data-path="lasso.html"><a href="lasso.html#code-lasso"><i class="fa fa-check"></i><b>11.7</b> Code: Lasso</a></li>
<li class="chapter" data-level="11.8" data-path="lasso.html"><a href="lasso.html#code-lasso-via-cyclic-gradient-descent"><i class="fa fa-check"></i><b>11.8</b> Code: Lasso via cyclic gradient descent</a></li>
<li class="chapter" data-level="11.9" data-path="lasso.html"><a href="lasso.html#code-parallel-lasso"><i class="fa fa-check"></i><b>11.9</b> Code: Parallel lasso</a></li>
<li class="chapter" data-level="11.10" data-path="lasso.html"><a href="lasso.html#references"><i class="fa fa-check"></i><b>11.10</b> References</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="tm.html"><a href="tm.html"><i class="fa fa-check"></i><b>12</b> Topic Models</a><ul>
<li class="chapter" data-level="12.1" data-path="tm.html"><a href="tm.html#variational-inference"><i class="fa fa-check"></i><b>12.1</b> Variational Inference</a><ul>
<li class="chapter" data-level="12.1.1" data-path="tm.html"><a href="tm.html#evidence-lower-bound-elbo"><i class="fa fa-check"></i><b>12.1.1</b> Evidence Lower Bound (ELBO)</a></li>
<li class="chapter" data-level="12.1.2" data-path="tm.html"><a href="tm.html#elbo-and-kl-divergence"><i class="fa fa-check"></i><b>12.1.2</b> ELBO and KL Divergence</a></li>
<li class="chapter" data-level="12.1.3" data-path="tm.html"><a href="tm.html#mean-field-method"><i class="fa fa-check"></i><b>12.1.3</b> Mean Field Method</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="tm.html"><a href="tm.html#lda"><i class="fa fa-check"></i><b>12.2</b> LDA</a><ul>
<li class="chapter" data-level="12.2.1" data-path="tm.html"><a href="tm.html#expectation-of-pthetaalpha"><i class="fa fa-check"></i><b>12.2.1</b> Expectation of p(<span class="math inline">\(\theta|\alpha)\)</span></a></li>
<li class="chapter" data-level="12.2.2" data-path="tm.html"><a href="tm.html#expectation-of-pztheta"><i class="fa fa-check"></i><b>12.2.2</b> Expectation of p(<span class="math inline">\(z|\theta)\)</span></a></li>
<li class="chapter" data-level="12.2.3" data-path="tm.html"><a href="tm.html#expectation-of-pwzbeta"><i class="fa fa-check"></i><b>12.2.3</b> Expectation of p(<span class="math inline">\(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.2.4" data-path="tm.html"><a href="tm.html#entropy-of-gamma-and-phi"><i class="fa fa-check"></i><b>12.2.4</b> Entropy of <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.2.5" data-path="tm.html"><a href="tm.html#complete-objective-function"><i class="fa fa-check"></i><b>12.2.5</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.2.6" data-path="tm.html"><a href="tm.html#parameter-optimization"><i class="fa fa-check"></i><b>12.2.6</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="tm.html"><a href="tm.html#supervised-lda"><i class="fa fa-check"></i><b>12.3</b> Supervised LDA</a><ul>
<li class="chapter" data-level="12.3.1" data-path="tm.html"><a href="tm.html#expectations-of-pthetaalpha-pztheta-and-pwzbeta"><i class="fa fa-check"></i><b>12.3.1</b> Expectations of <span class="math inline">\(p(\theta|\alpha)\)</span>, <span class="math inline">\(p(z|\theta)\)</span>, and <span class="math inline">\(p(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.3.2" data-path="tm.html"><a href="tm.html#expectation-of-pyzetasigma2"><i class="fa fa-check"></i><b>12.3.2</b> Expectation of <span class="math inline">\(p(y|z,\eta,\sigma^2)\)</span></a></li>
<li class="chapter" data-level="12.3.3" data-path="tm.html"><a href="tm.html#entropy-of-gamma-and-phi-1"><i class="fa fa-check"></i><b>12.3.3</b> Entropy of <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.3.4" data-path="tm.html"><a href="tm.html#complete-objective-function-1"><i class="fa fa-check"></i><b>12.3.4</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.3.5" data-path="tm.html"><a href="tm.html#parameter-optimization-1"><i class="fa fa-check"></i><b>12.3.5</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="tm.html"><a href="tm.html#the-correlated-topic-model"><i class="fa fa-check"></i><b>12.4</b> The Correlated Topic Model</a><ul>
<li class="chapter" data-level="12.4.1" data-path="tm.html"><a href="tm.html#multinomial-distribution-in-exponential-form"><i class="fa fa-check"></i><b>12.4.1</b> Multinomial Distribution in Exponential Form</a></li>
<li class="chapter" data-level="12.4.2" data-path="tm.html"><a href="tm.html#variational-em"><i class="fa fa-check"></i><b>12.4.2</b> Variational EM</a></li>
<li class="chapter" data-level="12.4.3" data-path="tm.html"><a href="tm.html#expectation-of-pwzbeta-1"><i class="fa fa-check"></i><b>12.4.3</b> Expectation of <span class="math inline">\(p(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.4.4" data-path="tm.html"><a href="tm.html#expectation-of-pzeta"><i class="fa fa-check"></i><b>12.4.4</b> Expectation of <span class="math inline">\(p(z|\eta)\)</span></a></li>
<li class="chapter" data-level="12.4.5" data-path="tm.html"><a href="tm.html#expectation-of-petamusigma"><i class="fa fa-check"></i><b>12.4.5</b> Expectation of <span class="math inline">\(p(\eta|\mu,\sigma)\)</span></a></li>
<li class="chapter" data-level="12.4.6" data-path="tm.html"><a href="tm.html#entropy-of-lambda-nu-and-phi"><i class="fa fa-check"></i><b>12.4.6</b> Entropy of <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\nu\)</span>, and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.4.7" data-path="tm.html"><a href="tm.html#complete-objective-function-2"><i class="fa fa-check"></i><b>12.4.7</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.4.8" data-path="tm.html"><a href="tm.html#parameter-optimization-2"><i class="fa fa-check"></i><b>12.4.8</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="tm.html"><a href="tm.html#dirichlet-distribution"><i class="fa fa-check"></i><b>12.5</b> Dirichlet Distribution</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="ml.html"><a href="ml.html"><i class="fa fa-check"></i><b>13</b> Machine Learning</a><ul>
<li class="chapter" data-level="13.1" data-path="ml.html"><a href="ml.html#cross-valdiation"><i class="fa fa-check"></i><b>13.1</b> Cross Valdiation</a><ul>
<li class="chapter" data-level="13.1.1" data-path="ml.html"><a href="ml.html#loocv"><i class="fa fa-check"></i><b>13.1.1</b> LOOCV</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="ml.html"><a href="ml.html#naive-bayes"><i class="fa fa-check"></i><b>13.2</b> Naive Bayes</a></li>
<li class="chapter" data-level="13.3" data-path="ml.html"><a href="ml.html#svm"><i class="fa fa-check"></i><b>13.3</b> SVM</a><ul>
<li class="chapter" data-level="13.3.1" data-path="ml.html"><a href="ml.html#manual-rbf-kernal"><i class="fa fa-check"></i><b>13.3.1</b> Manual rbf kernal</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="ml.html"><a href="ml.html#k-means"><i class="fa fa-check"></i><b>13.4</b> K-means</a></li>
<li class="chapter" data-level="13.5" data-path="ml.html"><a href="ml.html#gaussian-mixtures"><i class="fa fa-check"></i><b>13.5</b> Gaussian Mixtures</a></li>
<li class="chapter" data-level="13.6" data-path="ml.html"><a href="ml.html#pca"><i class="fa fa-check"></i><b>13.6</b> PCA</a></li>
<li class="chapter" data-level="13.7" data-path="ml.html"><a href="ml.html#viterbi-algorithm"><i class="fa fa-check"></i><b>13.7</b> Viterbi Algorithm</a></li>
<li class="chapter" data-level="13.8" data-path="ml.html"><a href="ml.html#gradient-descent"><i class="fa fa-check"></i><b>13.8</b> Gradient Descent</a><ul>
<li class="chapter" data-level="13.8.1" data-path="ml.html"><a href="ml.html#linear-regression"><i class="fa fa-check"></i><b>13.8.1</b> Linear Regression</a></li>
<li class="chapter" data-level="13.8.2" data-path="ml.html"><a href="ml.html#logistic-regression"><i class="fa fa-check"></i><b>13.8.2</b> Logistic Regression</a></li>
<li class="chapter" data-level="13.8.3" data-path="ml.html"><a href="ml.html#softmax-regression"><i class="fa fa-check"></i><b>13.8.3</b> Softmax regression</a></li>
</ul></li>
<li class="chapter" data-level="13.9" data-path="ml.html"><a href="ml.html#nonparametric-bayesian-processes"><i class="fa fa-check"></i><b>13.9</b> Nonparametric Bayesian Processes</a><ul>
<li class="chapter" data-level="13.9.1" data-path="ml.html"><a href="ml.html#chinese-restaurant"><i class="fa fa-check"></i><b>13.9.1</b> Chinese Restaurant</a></li>
<li class="chapter" data-level="13.9.2" data-path="ml.html"><a href="ml.html#polyas-urn"><i class="fa fa-check"></i><b>13.9.2</b> Polyas Urn</a></li>
<li class="chapter" data-level="13.9.3" data-path="ml.html"><a href="ml.html#stick-breaking"><i class="fa fa-check"></i><b>13.9.3</b> Stick Breaking</a></li>
</ul></li>
<li class="chapter" data-level="13.10" data-path="ml.html"><a href="ml.html#iteratively-reweighted-least-squares"><i class="fa fa-check"></i><b>13.10</b> Iteratively Reweighted Least Squares</a></li>
<li class="chapter" data-level="13.11" data-path="ml.html"><a href="ml.html#neural-network"><i class="fa fa-check"></i><b>13.11</b> Neural Network</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="microbiome.html"><a href="microbiome.html"><i class="fa fa-check"></i><b>14</b> Microbiome Data Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="microbiome.html"><a href="microbiome.html#loading-and-exploring-the-guerroro-negro-data"><i class="fa fa-check"></i><b>14.1</b> Loading and Exploring the Guerroro Negro Data</a></li>
<li class="chapter" data-level="14.2" data-path="microbiome.html"><a href="microbiome.html#statistical-analysis"><i class="fa fa-check"></i><b>14.2</b> Statistical Analysis</a><ul>
<li class="chapter" data-level="14.2.1" data-path="microbiome.html"><a href="microbiome.html#pca-1"><i class="fa fa-check"></i><b>14.2.1</b> PCA</a></li>
<li class="chapter" data-level="14.2.2" data-path="microbiome.html"><a href="microbiome.html#rda"><i class="fa fa-check"></i><b>14.2.2</b> RDA</a></li>
<li class="chapter" data-level="14.2.3" data-path="microbiome.html"><a href="microbiome.html#ca"><i class="fa fa-check"></i><b>14.2.3</b> CA</a></li>
<li class="chapter" data-level="14.2.4" data-path="microbiome.html"><a href="microbiome.html#cca"><i class="fa fa-check"></i><b>14.2.4</b> CCA</a></li>
<li class="chapter" data-level="14.2.5" data-path="microbiome.html"><a href="microbiome.html#pcoa"><i class="fa fa-check"></i><b>14.2.5</b> PCoA</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/sw1/" target="blank">Published by Stephen Woloszynek</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Worked Bioninformatics, Statistics, and Machine Learning Examples</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="qiime" class="section level1">
<h1><span class="header-section-number">Chapter 9</span> Qiime</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(Biostrings)</code></pre></div>
<p>First we need to use some qiime functions via the command line, so we’ll save them as variable names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">validate_mapping_file &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/validate_mapping_file.py&#39;</span>
split_libraries_fastq &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/split_libraries_fastq.py&#39;</span>
count_seqs &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/count_seqs.py&#39;</span>
extract_barcodes &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/extract_barcodes.py&#39;</span>
pick_closed_reference_otus &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/pick_closed_reference_otus.py&#39;</span>
core_diversity_analyses &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/core_diversity_analyses.py&#39;</span>
make_emperor &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/make_emperor.py&#39;</span>
biom &lt;-<span class="st"> &#39;/data/sw1/anaconda3/envs/qiime1/bin/biom&#39;</span></code></pre></div>
<div id="otu-picking-1" class="section level2">
<h2><span class="header-section-number">9.1</span> OTU Picking</h2>
<p>This will be an introduction to QIIME. We’ll start by going command by command, and then at the end will be a submission script for proteus. This should give you two flavors of running QIIME, one from within R, should you choose to install QIIME locally, and one via Proteus.</p>
<p>We’re going to use the dataset from the QIIME illumina tutorial, moving pictures, simply because it’s small and reliable. If you choose to install QIIME and worth through this locally, you can download the dataset by running the following command in bash, or you can simply go to the link and download it manually as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">svn checkout https:<span class="er">//</span>github.com/sw1/Bioinformatics/trunk/Data/moving_pictures<span class="er">}</span></code></pre></div>
<p>Once it’s downloaded, set the directly to a variable:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data_dir &lt;-<span class="st"> &#39;data/data_moving_pictures&#39;</span></code></pre></div>
<p>Now, we have a large FASTQ file with a ton of sequences:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">FASTQ &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(<span class="kw">file.path</span>(data_dir,<span class="st">&#39;forward_reads.fastq.gz&#39;</span>),<span class="dt">format=</span><span class="st">&#39;fastq&#39;</span>)
FASTQ</code></pre></div>
<pre><code>##   A DNAStringSet instance of length 302581
##          width seq                                          names               
##      [1]   152 TACGNAGGATCCGAGCGTTAT...GGCAGGGGGGGATTGGTGTG HWI-EAS440_0386:1...
##      [2]   152 CCCCNCAGCGGCAAAAATTAA...TGATGATTCCACTGCAACAA HWI-EAS440_0386:1...
##      [3]   152 TACGNAGGATCCGAGCGTTAT...GGCAGGGGGGGGGTTGGGGG HWI-EAS440_0386:1...
##      [4]   152 TACGNAGGATCCGAGCGTTAT...GGCAGGGGGGAGTTTGGGGG HWI-EAS440_0386:1...
##      [5]   152 TACGNAGGATCCGAGCGTTAT...GGCAGGCGGGATTCGTGGTG HWI-EAS440_0386:1...
##      ...   ... ...
## [302577]   152 NTGGCTGTTGGTTTCTCTGTG...TTCAGAATCAGAATGAGCCG HWI-EAS440_0386:6...
## [302578]   152 NACGTAGGTGGCAAGCGTTGT...GAAAGTGGAATTCCTAGTGA HWI-EAS440_0386:6...
## [302579]   152 NACGTAGGGTGCGAGCGTTAA...GGGAGGTAGAATTACACGTG HWI-EAS440_0386:6...
## [302580]   152 NACGTAGGGTGCGAGCGTTAA...GGGAGGTAGAACTCCACGTG HWI-EAS440_0386:6...
## [302581]   152 NACGTAGGTGGCAAGCGTTGT...GAGAGGTGGATTCATAGGAG HWI-EAS440_0386:6...</code></pre>
<p>Note the sequence header names; they have an interesting format, for examples</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">HWI-EAS440_0386:<span class="dv">1</span>:<span class="dv">23</span>:<span class="dv">17547</span>:<span class="dv">1423</span><span class="co">#0/1</span></code></pre></div>
<p>If you look at the figure below, you can see a detailed explaination of what this all means if you are interested, but it is information from illumina sequencing. Basically, when we sequence samples using illumina, we sequence multiple samples in a single flowcell lane to save money. These headers have some of that lane information encoded in them.</p>
<div class="figure">
<img src="figs/seq_id.png" alt="Seq id" />
<p class="caption">Seq id</p>
</div>
<p>Because we are mixing samples within a lane, we need a way to keep track of the samples, such that we know which sequence belongs to Frank and which belongs to Emily, despite many of their sequences sharing a lane. The way we keep track is through barcoding, where we add a short nucleotide sequence to the adapters used in PCR amplification. We therefore create sequence libraries, with each library receiving its own barcode, perform sequencing, and then parse these libraries using our barcodes to reassign sequences to our original samples.</p>
<p>Look at the metadata mapping file below. This contains all of the relevant sample information, as well as information associated with sequencing, particularly barcode information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MAP &lt;-<span class="st"> </span><span class="kw">read_delim</span>(<span class="kw">file.path</span>(data_dir,<span class="st">&#39;map.tsv&#39;</span>),<span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   `#SampleID` = col_character(),
##   BarcodeSequence = col_character(),
##   LinkerPrimerSequence = col_character(),
##   SampleType = col_character(),
##   Year = col_double(),
##   Month = col_double(),
##   Day = col_double(),
##   Subject = col_double(),
##   ReportedAntibioticUsage = col_character(),
##   DaysSinceExperimentStart = col_double(),
##   Description = col_character()
## )</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MAP[<span class="dv">1</span>:<span class="dv">5</span>,<span class="dv">1</span>:<span class="dv">5</span>]</code></pre></div>
<pre><code>## # A tibble: 5 x 5
##   `#SampleID` BarcodeSequence LinkerPrimerSequence SampleType  Year
##   &lt;chr&gt;       &lt;chr&gt;           &lt;chr&gt;                &lt;chr&gt;      &lt;dbl&gt;
## 1 L1S8        AGCTGACTAGTC    GTGCCAGCMGCCGCGGTAA  gut         2008
## 2 L1S140      ATGGCAGCTCTA    GTGCCAGCMGCCGCGGTAA  gut         2008
## 3 L1S57       ACACACTATGGC    GTGCCAGCMGCCGCGGTAA  gut         2009
## 4 L1S208      CTGAGATACGCG    GTGCCAGCMGCCGCGGTAA  gut         2009
## 5 L1S76       ACTACGTGTGGT    GTGCCAGCMGCCGCGGTAA  gut         2009</code></pre>
<p>We can see that each sample receives its own unique barcode:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nrow</span>(MAP[,<span class="dv">1</span>:<span class="dv">2</span>]) ==<span class="st"> </span><span class="kw">nrow</span>(<span class="kw">unique</span>(MAP[,<span class="dv">1</span>:<span class="dv">2</span>]))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Now take a look at the barcode file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BARCODE &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(<span class="kw">file.path</span>(data_dir,<span class="st">&#39;barcodes.fastq.gz&#39;</span>),<span class="dt">format=</span><span class="st">&#39;fastq&#39;</span>)
BARCODE</code></pre></div>
<pre><code>##   A DNAStringSet instance of length 302581
##          width seq                                          names               
##      [1]    12 ATGCAGCTCAGT                                 HWI-EAS440_0386:1...
##      [2]    12 CCCCTCAGCGGC                                 HWI-EAS440_0386:1...
##      [3]    12 GACGAGTCAGTC                                 HWI-EAS440_0386:1...
##      [4]    12 AGCAGTCGCGAT                                 HWI-EAS440_0386:1...
##      [5]    12 AGCACACCTACA                                 HWI-EAS440_0386:1...
##      ...   ... ...
## [302577]    12 ATGGCTGTTGGT                                 HWI-EAS440_0386:6...
## [302578]    12 ACACACTATGGC                                 HWI-EAS440_0386:6...
## [302579]    12 CAGCGGTGACAT                                 HWI-EAS440_0386:6...
## [302580]    12 GATCTTCAGTAC                                 HWI-EAS440_0386:6...
## [302581]    12 GACAGCGTTGAC                                 HWI-EAS440_0386:6...</code></pre>
<p>We can see that each barcode is assigned one of the header names we saw before. What we need to do now is break up this large fastq file that contains all of our sequence information into a new large file that has headers with unique <em>sample IDs</em> for each sequence, for example, Emily_1, Emily_2, and Emily_3 for 3 sequences belonging to Emily. We can easily do this in QIIME, but we first need to ensure that our mapping file is formatted correctly; otherwise, QIIME will not run. We can do that with the following command:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">system2</span>(validate_mapping_file,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-m&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;map.tsv&#39;</span>)))</code></pre></div>
<p>Now, let’s make some bad mapping files to see what happens. First, let’s simply remove the hashtag in the first column name:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MAP_bad1 &lt;-<span class="st"> </span>MAP
<span class="kw">colnames</span>(MAP_bad1)[<span class="dv">1</span>] &lt;-<span class="st"> &#39;SampleID&#39;</span>

tmp &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
<span class="kw">write_delim</span>(MAP_bad1,<span class="dt">path=</span>tmp,<span class="dt">delim=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)
<span class="kw">system2</span>(validate_mapping_file,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-m&#39;</span>,tmp))</code></pre></div>
<p>Now, look what happens if we don’t have a description column as the last column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MAP_bad2 &lt;-<span class="st"> </span>MAP
MAP_bad2 &lt;-<span class="st"> </span>MAP_bad2[,-<span class="kw">ncol</span>(MAP_bad2)]

tmp &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
<span class="kw">write_delim</span>(MAP_bad2,<span class="dt">path=</span>tmp,<span class="dt">delim=</span><span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)
<span class="kw">system2</span>(validate_mapping_file,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-m&#39;</span>,tmp))</code></pre></div>
<p>And lastly, what if we used separate with commas instead of tabs:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MAP_bad3 &lt;-<span class="st"> </span>MAP

tmp &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
<span class="kw">write_delim</span>(MAP_bad3,<span class="dt">path=</span>tmp,<span class="dt">delim=</span><span class="st">&#39;,&#39;</span>)
<span class="kw">system2</span>(validate_mapping_file,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-m&#39;</span>,tmp))</code></pre></div>
<p>Given that we have a good mapping file, let’s actually demultiplex our data. The following QIIME command performs demultiplexing and also the subsequent quality filtering. It also tosses out sequences that have poor matches to a given barcode. All of these parameters we can adjust.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">out &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
<span class="kw">system2</span>(validate_mapping_file,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-m&#39;</span>,tmp,
                                     <span class="st">&#39;-o&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;map_out&#39;</span>)))

<span class="kw">system2</span>(split_libraries_fastq,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-o&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;fastq_out_1&#39;</span>),
                                     <span class="st">&#39;-i&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;forward_reads.fastq.gz&#39;</span>),
                                     <span class="st">&#39;-b&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;barcodes.fastq.gz&#39;</span>),
                                     <span class="st">&#39;-m&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;map.tsv&#39;</span>)))</code></pre></div>
<p>Now, there are a few noteable filtering arguments:</p>
<ul>
<li>phred_quality_threshold</li>
<li>max_bad_run_length</li>
<li>min_per_read_length_fraction</li>
</ul>
<p>Phred score is defined as follows:</p>
<p><span class="math display">\[ 
Q=-10\log_{10}P 
\]</span> where P is the probability of a base-call error; hence,</p>
<p><span class="math display">\[
P=10^{-Q/10}
\]</span></p>
<p>The default <strong>phred_quality_threshold</strong> Q is 3, implying <span class="math inline">\(P=10^{-3/10}=.5\)</span>, or the probability of an incorrectly called base is about 50%. Setting it at 3 will flag any base call if it has a probability of being an error above this threshold. For the entire sequence, it will then look at <strong>max_bad_run_length</strong>, which checks how many <em>consecutive</em> flagged base calls are present. If this number is above a given threshold (the default is 3), then the sequence is truncated at that position. Finally, it checks <strong>min_per_read_length_fraction</strong>. This tosses any sequences that are shorter than a given length after truncation (the default is a length shorter than 75% of the original, unaltered read).</p>
<p>There is also a setting called <strong>phred offset</strong>, which you typically need not worry about (it’s automatically set), but is worth knowing what it represents:</p>
<div class="figure">
<img src="figs/phred_offset.png" alt="Phred offset" />
<p class="caption">Phred offset</p>
</div>
<p>I’d argue that something along the lines of Q=20 or greater makes more sense. Here, we’ll truncate at consecutive stretches of base calls where each call has 1% probability of being an error.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">system2</span>(split_libraries_fastq,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-o&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;fastq_out_2&#39;</span>),
                                     <span class="st">&#39;-i&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;forward_reads.fastq.gz&#39;</span>),
                                     <span class="st">&#39;-b&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;barcodes.fastq.gz&#39;</span>),
                                     <span class="st">&#39;-m&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;map.tsv&#39;</span>),
                                     <span class="st">&#39;-q&#39;</span>,<span class="st">&#39;20&#39;</span>))</code></pre></div>
<p>We’ll now perform OTU picking on our filtered, demultiplexed sequences. There are a ton of parameters we can adjust: <a href="http://qiime.org/scripts/pick_otus.html" class="uri">http://qiime.org/scripts/pick_otus.html</a>. To keep it somewhat simple, we’ll perform a particular kind of OTU picking called ‘closed.’ There are 3 OTU picking strategies: closed, open, and de novo.</p>
<p>Closed reference OTU picking is fast because it simply clusters sequences into OTUs based on reference sequences found in a database. If a sequence doesn’t match above a given similiary threshold to the database sequence, it is tossed away. It should probably be obvious if a given sequence isn’t found in the reference database, it too will be removed. De novo, on the other hand, can deal with novel sequences since it doesn’t use a lookup database. Instead, it clusters sequences by aligning them against one another, with sequences above a given similarity threshold clustered into OTUs. Lastly, open reference OTU picking first performed closed reference OTU picking, and then performs de novo on sequences that failed to match reference sequences in the lookup database.</p>
<p>To perform closed reference OTU picking, we need to specify our reference database (we’ll use GreenGenes) by setting the location of our reference sequences, our reference taxonomy. We also need to specify our picking parameters, which in this example, will consist of the OTU picking method (sortmerna), the number of threads for paralellization, and the similarity threshold (.97) for our sequences when we compare them to the GreenGenes database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ref_seqs &lt;-<span class="st"> &#39;/data/sw1/anaconda2/lib/python2.7/site-packages/qiime_default_reference/gg_13_8_otus/rep_set/97_otus.fasta&#39;</span>
ref_tax &lt;-<span class="st"> &#39;/data/sw1/anaconda2/lib/python2.7/site-packages/qiime_default_reference/gg_13_8_otus/taxonomy/97_otu_taxonomy.txt&#39;</span>

params &lt;-<span class="st"> </span><span class="kw">tempfile</span>(<span class="dt">fileext=</span><span class="st">&#39;.txt&#39;</span>)
<span class="kw">write_lines</span>(<span class="kw">c</span>(<span class="st">&#39;pick_otus:otu_picking_method sortmerna&#39;</span>,
              <span class="st">&#39;pick_otus:threads 4&#39;</span>,
              <span class="st">&#39;pick_otus:similarity: 0.97&#39;</span>),<span class="dt">path=</span>params)</code></pre></div>
<p>And now we run the closed reference otu picking command:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">system2</span>(pick_closed_reference_otus,<span class="dt">args=</span><span class="kw">c</span>(<span class="st">&#39;-i&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;fastq_out_2&#39;</span>,<span class="st">&#39;seqs.fna&#39;</span>),
                                          <span class="st">&#39;-o&#39;</span>,<span class="kw">file.path</span>(data_dir,<span class="st">&#39;fastq_out_2&#39;</span>,<span class="st">&#39;picked_otus_sortmerna&#39;</span>),
                                          <span class="st">&#39;-r&#39;</span>,ref_seqs,
                                          <span class="st">&#39;-t&#39;</span>,ref_tax,
                                          <span class="st">&#39;-p&#39;</span>,params))</code></pre></div>
<p>This was all done in R, which is probably unnecessary for proteus. You can submit an Rscript, but since there is no real looping and need for immediate downstream analysis, it’s better off just submitting a bash script. We’ll remove the system2 functions and the like, add the submission parameters, and write the following script for proteus:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#!/bin/bash</span>
<span class="co">#$ -S /bin/bash</span>
<span class="co">#$ -cwd</span>
<span class="co">#$ -j y</span>
<span class="co">#$ -M user_name@email.edu</span>
<span class="co">#$ -l h_rt=01:00:00</span>
<span class="co">#$ -P namePrj</span>
<span class="co">#$ -pe shm 16</span>
<span class="co">#$ -l mem_free=12G</span>
<span class="co">#$ -l h_vmem=16G</span>
<span class="co">#$ -q all.q</span>

. /etc/profile.d/modules.sh
module load shared
module load proteus
module load sge/univa
module load gcc/<span class="fl">4.8.1</span>
module load qiime/gcc/<span class="dv">64</span>/<span class="fl">1.9.1</span>

export ref_seqs=<span class="er">/</span>mnt/HA/opt/qiime/gcc/<span class="dv">64</span>/<span class="fl">1.9.1</span>/lib/python2<span class="fl">.7</span>/site-packages/qiime_default_reference/gg_13_8_otus/rep_set/97_otus.fasta
export ref_tax=<span class="er">/</span>mnt/HA/opt/qiime/gcc/<span class="dv">64</span>/<span class="fl">1.9.1</span>/lib/python2<span class="fl">.7</span>/site-packages/qiime_default_reference/gg_13_8_otus/taxonomy/97_otu_taxonomy.txt

data_dir=<span class="er">/</span>home/user_name/genStats/moving_pictures
work_dir=<span class="er">/</span>scratch/user_name/moving_pictures

mkdir -p $work_dir

cp $data_dir/<span class="er">*</span><span class="st"> </span><span class="er">$</span>work_dir

out_dir=<span class="er">$</span>work_dir/fastq_out
seqs=<span class="er">$</span>work_dir/forward_reads.fastq.gz
bc=<span class="er">$</span>work_dir/barcodes.fastq.gz
map=<span class="er">$</span>work_dir/map.tsv

split_libraries_fastq.py -o $out_dir -i $seqs -b $bc -m $map -q <span class="dv">20</span>

params=<span class="er">$</span>work_dir/sortmerna_pick_params.txt

printf <span class="st">&quot;pick_otus:otu_picking_method sortmerna</span><span class="ch">\n</span><span class="st">pick_otus:threads 16</span><span class="ch">\n</span><span class="st">pick_otus:similarity 0.97&quot;</span> &gt;<span class="st"> </span><span class="er">$</span>params

pick_closed_reference_otus.py -i $out_dir/seqs.fna -o $out_dir/picked_otus -r $ref_seqs -t $ref_tax -p $params

mv $out_dir $data_dir
rm -r $work_dir/<span class="er">*</span>
<span class="st">  </span>
exit <span class="dv">0</span></code></pre></div>
</div>
<div id="summarizing-our-results" class="section level2">
<h2><span class="header-section-number">9.2</span> Summarizing Our Results</h2>
<p>Now that we have our OTU table, in biom format, we can summarize our results using some QIIME commands. First, let’s look at the mapping file to see what sample features we’d like to focus on:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">View</span>(MAP)</code></pre></div>
<p>Now, we’ll write a submission script to get analyses output.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#!/bin/bash</span>
<span class="co">#$ -S /bin/bash</span>
<span class="co">#$ -cwd</span>
<span class="co">#$ -j y</span>
<span class="co">#$ -M user_name@email.edu</span>
<span class="co">#$ -l h_rt=01:00:00</span>
<span class="co">#$ -P namePrj</span>
<span class="co">#$ -l mem_free=12G</span>
<span class="co">#$ -l h_vmem=16G</span>
<span class="co">#$ -q all.q</span>

. /etc/profile.d/modules.sh
module load shared
module load proteus
module load sge/univa
module load gcc/<span class="fl">4.8.1</span>
module load qiime/gcc/<span class="dv">64</span>/<span class="fl">1.9.1</span>

export ref_seqs=<span class="er">/</span>mnt/HA/opt/qiime/gcc/<span class="dv">64</span>/<span class="fl">1.9.1</span>/lib/python2<span class="fl">.7</span>/site-packages/qiime_default_reference/gg_13_8_otus/rep_set/97_otus.fasta
export ref_tax=<span class="er">/</span>mnt/HA/opt/qiime/gcc/<span class="dv">64</span>/<span class="fl">1.9.1</span>/lib/python2<span class="fl">.7</span>/site-packages/qiime_default_reference/gg_13_8_otus/taxonomy/97_otu_taxonomy.txt

data_dir=<span class="er">/</span>home/user_name/genStats/moving_pictures
work_dir=<span class="er">/</span>scratch/user_name/moving_pictures

mkdir -p $work_dir

cp $data_dir/fastq_out/picked_otus/97_otus.tree $work_dir
cp $data_dir/fastq_out/picked_otus/otu_table.biom $work_dir
cp $data_dir/map.tsv $work_dir

otus=<span class="er">$</span>work_dir/otu_table.biom
tree=<span class="er">$</span>work_dir/97_otus.tree
map=<span class="er">$</span>work_dir/map.tsv

analysis_dir=<span class="er">$</span>work_dir/analysis

mkdir -p $analysis_dir

biom summarize-table -i $otus &gt;<span class="st"> </span><span class="er">$</span>analysis_dir/sampling_depth.dat

core_diversity_analyses.py -o $analysis_dir/diversity -i $otus -m $map -t $tree -c SampleType -e <span class="dv">1000</span> --recover_from_failure

mv $analysis_dir $data_dir
rm -r $work_dir/<span class="er">*</span>
<span class="st">  </span>
exit <span class="dv">0</span></code></pre></div>
<p>Or we can calculate the metrics more directly:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#!/bin/bash</span>
<span class="co">#$ -S /bin/bash</span>
<span class="co">#$ -cwd</span>
<span class="co">#$ -j y</span>
<span class="co">#$ -M user_name@email.edu</span>
<span class="co">#$ -l h_rt=01:00:00</span>
<span class="co">#$ -P namePrj</span>
<span class="co">#$ -l mem_free=12G</span>
<span class="co">#$ -l h_vmem=16G</span>
<span class="co">#$ -q all.q</span>

. /etc/profile.d/modules.sh
module load shared
module load proteus
module load sge/univa
module load gcc/<span class="fl">4.8.1</span>
module load qiime/gcc/<span class="dv">64</span>/<span class="fl">1.9.1</span>

data_dir=<span class="er">/</span>home/user_name/genStats/moving_pictures
work_dir=<span class="er">/</span>scratch/user_name/moving_pictures/

mkdir -p $work_dir

cp $data_dir/fastq_out/picked_otus/97_otus.tree $work_dir
cp $data_dir/fastq_out/picked_otus/otu_table.biom $work_dir
cp $data_dir/map.tsv $work_dir

otus=<span class="er">$</span>work_dir/otu_table.biom
tree=<span class="er">$</span>work_dir/97_otus.tree
map=<span class="er">$</span>work_dir/map.tsv


met_dir=<span class="er">$</span>work_dir/metrics


alpha_diversity.py -i $otus -t $tree -o $met_dir/alpha.txt

beta_diversity.py -i $otus -o $met_dir/beta -t $tree -m unweighted_unifrac
beta_diversity.py -i $otus -o $met_dir/beta -t $tree -m weighted_unifrac

principal_coordinates.py -i $met_dir/beta -o $met_dir/pcoa

make_2d_plots.py -i $met_dir/pcoa/pcoa_unweighted_unifrac_otu_table.txt -m $map -b SampleType -o $met_dir/pcoa/figures
make_2d_plots.py -i $met_dir/pcoa/pcoa_weighted_unifrac_otu_table.txt -m $map -b SampleType -o $met_dir/pcoa/figures

transform_coordinate_matrices.py -i $met_dir/pcoa/pcoa_unweighted_unifrac_otu_table.txt,$met_dir/pcoa/pcoa_weighted_unifrac_otu_table.txt -r <span class="dv">999</span> -o $met_dir/procrustes/

make_emperor.py -c -i $met_dir/procrustes -o $met_dir/procrustes/figures -m $map --custom_axes DaysSinceExperimentStart

mv $met_dir $data_dir

exit <span class="dv">0</span></code></pre></div>
</div>
<div id="loading-qiime-results-into-phyloseq" class="section level2">
<h2><span class="header-section-number">9.3</span> Loading QIIME Results into Phyloseq</h2>
<p>We have a ton of output from QIIME, but we could have just as easily loaded the biom table into R, specifically phyloseq. We can do this like so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(phyloseq)</code></pre></div>
<pre><code>## 
## Attaching package: &#39;phyloseq&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     distance</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">biom_path &lt;-<span class="st"> </span><span class="kw">file.path</span>(data_dir,<span class="st">&#39;fastq_out_2&#39;</span>,<span class="st">&#39;picked_otus_sortmerna&#39;</span>,<span class="st">&#39;otu_table.biom&#39;</span>)
tree_path &lt;-<span class="st"> </span><span class="kw">file.path</span>(data_dir,<span class="st">&#39;fastq_out_2&#39;</span>,<span class="st">&#39;picked_otus_sortmerna&#39;</span>,<span class="st">&#39;97_otus.tree&#39;</span>)

BIOM &lt;-<span class="st"> </span><span class="kw">import_biom</span>(biom_path)</code></pre></div>
<pre><code>## Warning in strsplit(conditionMessage(e), &quot;\n&quot;): input string 1 is invalid in
## this locale</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">META &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(MAP)
<span class="kw">rownames</span>(META) &lt;-<span class="st"> </span>META$<span class="st">`</span><span class="dt">#SampleID</span><span class="st">`</span>
PS &lt;-<span class="st"> </span><span class="kw">merge_phyloseq</span>(BIOM,<span class="kw">sample_data</span>(META))
PS</code></pre></div>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 3409 taxa and 34 samples ]
## sample_data() Sample Data:       [ 34 samples by 11 sample variables ]
## tax_table()   Taxonomy Table:    [ 3409 taxa by 7 taxonomic ranks ]</code></pre>
<p>And then perform similar analyses:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_ordination</span>(PS,<span class="kw">ordinate</span>(PS,<span class="dt">method=</span><span class="st">&#39;MDS&#39;</span>,<span class="dt">distance=</span><span class="st">&#39;bray&#39;</span>),<span class="dt">type=</span><span class="st">&#39;samples&#39;</span>,<span class="dt">color=</span><span class="st">&#39;SampleType&#39;</span>)</code></pre></div>
<p><img src="08_qiime_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dada.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="multcomp.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
