<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Retrieving Projects | Worked Bioninformatics, Statistics, and Machine Learning Examples</title>
  <meta name="description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  <meta name="generator" content="bookdown 0.17 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Retrieving Projects | Worked Bioninformatics, Statistics, and Machine Learning Examples" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Retrieving Projects | Worked Bioninformatics, Statistics, and Machine Learning Examples" />
  
  <meta name="twitter:description" content="Code, notes, lectures, and research that should be helpful for learning purposes." />
  

<meta name="author" content="Stephen Woloszynek" />


<meta name="date" content="2020-02-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="biocond.html"/>
<link rel="next" href="phyloseq.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="dynprog.html"><a href="dynprog.html"><i class="fa fa-check"></i><b>2</b> Dynamic Programming</a><ul>
<li class="chapter" data-level="2.1" data-path="dynprog.html"><a href="dynprog.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="dynprog.html"><a href="dynprog.html#rod-cutting"><i class="fa fa-check"></i><b>2.2</b> Rod cutting</a></li>
<li class="chapter" data-level="2.3" data-path="dynprog.html"><a href="dynprog.html#fibonacci-rabbits"><i class="fa fa-check"></i><b>2.3</b> Fibonacci rabbits</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="align.html"><a href="align.html"><i class="fa fa-check"></i><b>3</b> Alignment</a><ul>
<li class="chapter" data-level="3.1" data-path="align.html"><a href="align.html#longest-common-subsequence"><i class="fa fa-check"></i><b>3.1</b> Longest Common Subsequence</a></li>
<li class="chapter" data-level="3.2" data-path="align.html"><a href="align.html#global-alignment"><i class="fa fa-check"></i><b>3.2</b> Global Alignment</a></li>
<li class="chapter" data-level="3.3" data-path="align.html"><a href="align.html#local-alignment"><i class="fa fa-check"></i><b>3.3</b> Local Alignment</a></li>
<li class="chapter" data-level="3.4" data-path="align.html"><a href="align.html#local-alignment-homework"><i class="fa fa-check"></i><b>3.4</b> Local Alignment: Homework</a></li>
<li class="chapter" data-level="3.5" data-path="align.html"><a href="align.html#global-alignment-code-r"><i class="fa fa-check"></i><b>3.5</b> Global Alignment Code (R)</a></li>
<li class="chapter" data-level="3.6" data-path="align.html"><a href="align.html#global-alignment-code-python"><i class="fa fa-check"></i><b>3.6</b> Global Alignment Code (Python)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="alignalg.html"><a href="alignalg.html"><i class="fa fa-check"></i><b>4</b> Alignment Algorithms</a><ul>
<li class="chapter" data-level="4.1" data-path="alignalg.html"><a href="alignalg.html#longest-common-subsequence-1"><i class="fa fa-check"></i><b>4.1</b> Longest Common Subsequence</a></li>
<li class="chapter" data-level="4.2" data-path="alignalg.html"><a href="alignalg.html#global-alignment-r"><i class="fa fa-check"></i><b>4.2</b> Global Alignment (R)</a></li>
<li class="chapter" data-level="4.3" data-path="alignalg.html"><a href="alignalg.html#global-alignment-python"><i class="fa fa-check"></i><b>4.3</b> Global Alignment (Python)</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="biocond.html"><a href="biocond.html"><i class="fa fa-check"></i><b>5</b> Bioconductor</a><ul>
<li class="chapter" data-level="5.0.1" data-path="biocond.html"><a href="biocond.html#loading-fasta-files"><i class="fa fa-check"></i><b>5.0.1</b> Loading FASTA Files</a></li>
<li class="chapter" data-level="5.0.2" data-path="biocond.html"><a href="biocond.html#creating-sequence-sets"><i class="fa fa-check"></i><b>5.0.2</b> Creating Sequence Sets</a></li>
<li class="chapter" data-level="5.0.3" data-path="biocond.html"><a href="biocond.html#sample-metadata"><i class="fa fa-check"></i><b>5.0.3</b> Sample Metadata</a></li>
<li class="chapter" data-level="5.1" data-path="biocond.html"><a href="biocond.html#creating-gc-functions"><i class="fa fa-check"></i><b>5.1</b> Creating GC Functions</a></li>
<li class="chapter" data-level="5.2" data-path="biocond.html"><a href="biocond.html#ncbi-esearch"><i class="fa fa-check"></i><b>5.2</b> NCBI ESearch</a></li>
<li class="chapter" data-level="5.3" data-path="biocond.html"><a href="biocond.html#cds"><i class="fa fa-check"></i><b>5.3</b> CDS</a></li>
<li class="chapter" data-level="5.4" data-path="biocond.html"><a href="biocond.html#whole-genomes"><i class="fa fa-check"></i><b>5.4</b> Whole Genomes</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="sra.html"><a href="sra.html"><i class="fa fa-check"></i><b>6</b> Retrieving Projects</a><ul>
<li class="chapter" data-level="6.0.1" data-path="sra.html"><a href="sra.html#fastq-dump-for-paired-end-reads"><i class="fa fa-check"></i><b>6.0.1</b> Fastq Dump for Paired End Reads</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="phyloseq.html"><a href="phyloseq.html"><i class="fa fa-check"></i><b>7</b> Phyloseq</a></li>
<li class="chapter" data-level="8" data-path="dada.html"><a href="dada.html"><i class="fa fa-check"></i><b>8</b> Dada2</a><ul>
<li class="chapter" data-level="8.1" data-path="dada.html"><a href="dada.html#fastq-prep"><i class="fa fa-check"></i><b>8.1</b> FASTQ Prep</a></li>
<li class="chapter" data-level="8.2" data-path="dada.html"><a href="dada.html#otu-picking"><i class="fa fa-check"></i><b>8.2</b> OTU Picking</a></li>
<li class="chapter" data-level="8.3" data-path="dada.html"><a href="dada.html#running-dada2-on-proteus"><i class="fa fa-check"></i><b>8.3</b> Running Dada2 on Proteus</a><ul>
<li class="chapter" data-level="8.3.1" data-path="dada.html"><a href="dada.html#method-1-using-namegrp-shared-r-library"><i class="fa fa-check"></i><b>8.3.1</b> Method 1: Using nameGrp Shared R Library</a></li>
<li class="chapter" data-level="8.3.2" data-path="dada.html"><a href="dada.html#method-2-creating-a-local-library"><i class="fa fa-check"></i><b>8.3.2</b> Method 2: Creating a Local Library</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="qiime.html"><a href="qiime.html"><i class="fa fa-check"></i><b>9</b> Qiime</a><ul>
<li class="chapter" data-level="9.1" data-path="qiime.html"><a href="qiime.html#otu-picking-1"><i class="fa fa-check"></i><b>9.1</b> OTU Picking</a></li>
<li class="chapter" data-level="9.2" data-path="qiime.html"><a href="qiime.html#summarizing-our-results"><i class="fa fa-check"></i><b>9.2</b> Summarizing Our Results</a></li>
<li class="chapter" data-level="9.3" data-path="qiime.html"><a href="qiime.html#loading-qiime-results-into-phyloseq"><i class="fa fa-check"></i><b>9.3</b> Loading QIIME Results into Phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="multcomp.html"><a href="multcomp.html"><i class="fa fa-check"></i><b>10</b> Multiple Comparisons</a><ul>
<li class="chapter" data-level="10.1" data-path="multcomp.html"><a href="multcomp.html#hypothesis-testing-and-power"><i class="fa fa-check"></i><b>10.1</b> Hypothesis Testing and Power</a></li>
<li class="chapter" data-level="10.2" data-path="multcomp.html"><a href="multcomp.html#multiple-comparisons"><i class="fa fa-check"></i><b>10.2</b> Multiple Comparisons</a><ul>
<li class="chapter" data-level="10.2.1" data-path="multcomp.html"><a href="multcomp.html#bonferroni"><i class="fa fa-check"></i><b>10.2.1</b> Bonferroni</a></li>
<li class="chapter" data-level="10.2.2" data-path="multcomp.html"><a href="multcomp.html#false-discovery-rate"><i class="fa fa-check"></i><b>10.2.2</b> False Discovery Rate</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="lasso.html"><a href="lasso.html"><i class="fa fa-check"></i><b>11</b> Lasso</a><ul>
<li class="chapter" data-level="11.1" data-path="lasso.html"><a href="lasso.html#big-data-and-feature-selection"><i class="fa fa-check"></i><b>11.1</b> Big Data and Feature Selection</a></li>
<li class="chapter" data-level="11.2" data-path="lasso.html"><a href="lasso.html#lasso-regression"><i class="fa fa-check"></i><b>11.2</b> Lasso Regression</a></li>
<li class="chapter" data-level="11.3" data-path="lasso.html"><a href="lasso.html#multivariate-lasso-regression"><i class="fa fa-check"></i><b>11.3</b> Multivariate Lasso Regression</a></li>
<li class="chapter" data-level="11.4" data-path="lasso.html"><a href="lasso.html#parallel-coordinate-descent"><i class="fa fa-check"></i><b>11.4</b> Parallel Coordinate Descent</a></li>
<li class="chapter" data-level="11.5" data-path="lasso.html"><a href="lasso.html#simulations"><i class="fa fa-check"></i><b>11.5</b> Simulations</a><ul>
<li class="chapter" data-level="11.5.1" data-path="lasso.html"><a href="lasso.html#times-1000-matrix-5-target-coefficients"><i class="fa fa-check"></i><b>11.5.1</b> <span class="math inline">\(50 \times 1000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.2" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-5-target-coefficients"><i class="fa fa-check"></i><b>11.5.2</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.3" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-5-target-coefficients-1"><i class="fa fa-check"></i><b>11.5.3</b> <span class="math inline">\(200 \times 10000\)</span> Matrix | 5 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.4" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-15-target-coefficients"><i class="fa fa-check"></i><b>11.5.4</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 15 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.5" data-path="lasso.html"><a href="lasso.html#times-10000-matrix-30-target-coefficients"><i class="fa fa-check"></i><b>11.5.5</b> <span class="math inline">\(50 \times 10000\)</span> Matrix | 30 Target Coefficients</a></li>
<li class="chapter" data-level="11.5.6" data-path="lasso.html"><a href="lasso.html#scaling-the-coeficients"><i class="fa fa-check"></i><b>11.5.6</b> Scaling the coeficients</a></li>
<li class="chapter" data-level="11.5.7" data-path="lasso.html"><a href="lasso.html#comparison"><i class="fa fa-check"></i><b>11.5.7</b> Comparison</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="lasso.html"><a href="lasso.html#conclusion"><i class="fa fa-check"></i><b>11.6</b> Conclusion</a></li>
<li class="chapter" data-level="11.7" data-path="lasso.html"><a href="lasso.html#code-lasso"><i class="fa fa-check"></i><b>11.7</b> Code: Lasso</a></li>
<li class="chapter" data-level="11.8" data-path="lasso.html"><a href="lasso.html#code-lasso-via-cyclic-gradient-descent"><i class="fa fa-check"></i><b>11.8</b> Code: Lasso via cyclic gradient descent</a></li>
<li class="chapter" data-level="11.9" data-path="lasso.html"><a href="lasso.html#code-parallel-lasso"><i class="fa fa-check"></i><b>11.9</b> Code: Parallel lasso</a></li>
<li class="chapter" data-level="11.10" data-path="lasso.html"><a href="lasso.html#references"><i class="fa fa-check"></i><b>11.10</b> References</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="tm.html"><a href="tm.html"><i class="fa fa-check"></i><b>12</b> Topic Models</a><ul>
<li class="chapter" data-level="12.1" data-path="tm.html"><a href="tm.html#variational-inference"><i class="fa fa-check"></i><b>12.1</b> Variational Inference</a><ul>
<li class="chapter" data-level="12.1.1" data-path="tm.html"><a href="tm.html#evidence-lower-bound-elbo"><i class="fa fa-check"></i><b>12.1.1</b> Evidence Lower Bound (ELBO)</a></li>
<li class="chapter" data-level="12.1.2" data-path="tm.html"><a href="tm.html#elbo-and-kl-divergence"><i class="fa fa-check"></i><b>12.1.2</b> ELBO and KL Divergence</a></li>
<li class="chapter" data-level="12.1.3" data-path="tm.html"><a href="tm.html#mean-field-method"><i class="fa fa-check"></i><b>12.1.3</b> Mean Field Method</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="tm.html"><a href="tm.html#lda"><i class="fa fa-check"></i><b>12.2</b> LDA</a><ul>
<li class="chapter" data-level="12.2.1" data-path="tm.html"><a href="tm.html#expectation-of-pthetaalpha"><i class="fa fa-check"></i><b>12.2.1</b> Expectation of p(<span class="math inline">\(\theta|\alpha)\)</span></a></li>
<li class="chapter" data-level="12.2.2" data-path="tm.html"><a href="tm.html#expectation-of-pztheta"><i class="fa fa-check"></i><b>12.2.2</b> Expectation of p(<span class="math inline">\(z|\theta)\)</span></a></li>
<li class="chapter" data-level="12.2.3" data-path="tm.html"><a href="tm.html#expectation-of-pwzbeta"><i class="fa fa-check"></i><b>12.2.3</b> Expectation of p(<span class="math inline">\(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.2.4" data-path="tm.html"><a href="tm.html#entropy-of-gamma-and-phi"><i class="fa fa-check"></i><b>12.2.4</b> Entropy of <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.2.5" data-path="tm.html"><a href="tm.html#complete-objective-function"><i class="fa fa-check"></i><b>12.2.5</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.2.6" data-path="tm.html"><a href="tm.html#parameter-optimization"><i class="fa fa-check"></i><b>12.2.6</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="tm.html"><a href="tm.html#supervised-lda"><i class="fa fa-check"></i><b>12.3</b> Supervised LDA</a><ul>
<li class="chapter" data-level="12.3.1" data-path="tm.html"><a href="tm.html#expectations-of-pthetaalpha-pztheta-and-pwzbeta"><i class="fa fa-check"></i><b>12.3.1</b> Expectations of <span class="math inline">\(p(\theta|\alpha)\)</span>, <span class="math inline">\(p(z|\theta)\)</span>, and <span class="math inline">\(p(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.3.2" data-path="tm.html"><a href="tm.html#expectation-of-pyzetasigma2"><i class="fa fa-check"></i><b>12.3.2</b> Expectation of <span class="math inline">\(p(y|z,\eta,\sigma^2)\)</span></a></li>
<li class="chapter" data-level="12.3.3" data-path="tm.html"><a href="tm.html#entropy-of-gamma-and-phi-1"><i class="fa fa-check"></i><b>12.3.3</b> Entropy of <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.3.4" data-path="tm.html"><a href="tm.html#complete-objective-function-1"><i class="fa fa-check"></i><b>12.3.4</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.3.5" data-path="tm.html"><a href="tm.html#parameter-optimization-1"><i class="fa fa-check"></i><b>12.3.5</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="tm.html"><a href="tm.html#the-correlated-topic-model"><i class="fa fa-check"></i><b>12.4</b> The Correlated Topic Model</a><ul>
<li class="chapter" data-level="12.4.1" data-path="tm.html"><a href="tm.html#multinomial-distribution-in-exponential-form"><i class="fa fa-check"></i><b>12.4.1</b> Multinomial Distribution in Exponential Form</a></li>
<li class="chapter" data-level="12.4.2" data-path="tm.html"><a href="tm.html#variational-em"><i class="fa fa-check"></i><b>12.4.2</b> Variational EM</a></li>
<li class="chapter" data-level="12.4.3" data-path="tm.html"><a href="tm.html#expectation-of-pwzbeta-1"><i class="fa fa-check"></i><b>12.4.3</b> Expectation of <span class="math inline">\(p(w|z,\beta)\)</span></a></li>
<li class="chapter" data-level="12.4.4" data-path="tm.html"><a href="tm.html#expectation-of-pzeta"><i class="fa fa-check"></i><b>12.4.4</b> Expectation of <span class="math inline">\(p(z|\eta)\)</span></a></li>
<li class="chapter" data-level="12.4.5" data-path="tm.html"><a href="tm.html#expectation-of-petamusigma"><i class="fa fa-check"></i><b>12.4.5</b> Expectation of <span class="math inline">\(p(\eta|\mu,\sigma)\)</span></a></li>
<li class="chapter" data-level="12.4.6" data-path="tm.html"><a href="tm.html#entropy-of-lambda-nu-and-phi"><i class="fa fa-check"></i><b>12.4.6</b> Entropy of <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\nu\)</span>, and <span class="math inline">\(\phi\)</span></a></li>
<li class="chapter" data-level="12.4.7" data-path="tm.html"><a href="tm.html#complete-objective-function-2"><i class="fa fa-check"></i><b>12.4.7</b> Complete Objective Function</a></li>
<li class="chapter" data-level="12.4.8" data-path="tm.html"><a href="tm.html#parameter-optimization-2"><i class="fa fa-check"></i><b>12.4.8</b> Parameter Optimization</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="tm.html"><a href="tm.html#dirichlet-distribution"><i class="fa fa-check"></i><b>12.5</b> Dirichlet Distribution</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="ml.html"><a href="ml.html"><i class="fa fa-check"></i><b>13</b> Machine Learning</a><ul>
<li class="chapter" data-level="13.1" data-path="ml.html"><a href="ml.html#cross-valdiation"><i class="fa fa-check"></i><b>13.1</b> Cross Valdiation</a><ul>
<li class="chapter" data-level="13.1.1" data-path="ml.html"><a href="ml.html#loocv"><i class="fa fa-check"></i><b>13.1.1</b> LOOCV</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="ml.html"><a href="ml.html#naive-bayes"><i class="fa fa-check"></i><b>13.2</b> Naive Bayes</a></li>
<li class="chapter" data-level="13.3" data-path="ml.html"><a href="ml.html#svm"><i class="fa fa-check"></i><b>13.3</b> SVM</a><ul>
<li class="chapter" data-level="13.3.1" data-path="ml.html"><a href="ml.html#manual-rbf-kernal"><i class="fa fa-check"></i><b>13.3.1</b> Manual rbf kernal</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="ml.html"><a href="ml.html#k-means"><i class="fa fa-check"></i><b>13.4</b> K-means</a></li>
<li class="chapter" data-level="13.5" data-path="ml.html"><a href="ml.html#gaussian-mixtures"><i class="fa fa-check"></i><b>13.5</b> Gaussian Mixtures</a></li>
<li class="chapter" data-level="13.6" data-path="ml.html"><a href="ml.html#pca"><i class="fa fa-check"></i><b>13.6</b> PCA</a></li>
<li class="chapter" data-level="13.7" data-path="ml.html"><a href="ml.html#viterbi-algorithm"><i class="fa fa-check"></i><b>13.7</b> Viterbi Algorithm</a></li>
<li class="chapter" data-level="13.8" data-path="ml.html"><a href="ml.html#gradient-descent"><i class="fa fa-check"></i><b>13.8</b> Gradient Descent</a><ul>
<li class="chapter" data-level="13.8.1" data-path="ml.html"><a href="ml.html#linear-regression"><i class="fa fa-check"></i><b>13.8.1</b> Linear Regression</a></li>
<li class="chapter" data-level="13.8.2" data-path="ml.html"><a href="ml.html#logistic-regression"><i class="fa fa-check"></i><b>13.8.2</b> Logistic Regression</a></li>
<li class="chapter" data-level="13.8.3" data-path="ml.html"><a href="ml.html#softmax-regression"><i class="fa fa-check"></i><b>13.8.3</b> Softmax regression</a></li>
</ul></li>
<li class="chapter" data-level="13.9" data-path="ml.html"><a href="ml.html#nonparametric-bayesian-processes"><i class="fa fa-check"></i><b>13.9</b> Nonparametric Bayesian Processes</a><ul>
<li class="chapter" data-level="13.9.1" data-path="ml.html"><a href="ml.html#chinese-restaurant"><i class="fa fa-check"></i><b>13.9.1</b> Chinese Restaurant</a></li>
<li class="chapter" data-level="13.9.2" data-path="ml.html"><a href="ml.html#polyas-urn"><i class="fa fa-check"></i><b>13.9.2</b> Polyas Urn</a></li>
<li class="chapter" data-level="13.9.3" data-path="ml.html"><a href="ml.html#stick-breaking"><i class="fa fa-check"></i><b>13.9.3</b> Stick Breaking</a></li>
</ul></li>
<li class="chapter" data-level="13.10" data-path="ml.html"><a href="ml.html#iteratively-reweighted-least-squares"><i class="fa fa-check"></i><b>13.10</b> Iteratively Reweighted Least Squares</a></li>
<li class="chapter" data-level="13.11" data-path="ml.html"><a href="ml.html#neural-netork"><i class="fa fa-check"></i><b>13.11</b> Neural Netork</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/sw1/" target="blank">Published by Stephen Woloszynek</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Worked Bioninformatics, Statistics, and Machine Learning Examples</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="sra" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Retrieving Projects</h1>
<p>Many of your projects will involve working with publically available data from published research. The data itself is typicaly uploaded to a data repository like NCBI or EMBL/EBI. The files are deposited as Sequence Read Archive (SRA) files and are often searchable as bioprojects. Accessing these files can be accomplished using Bioconductor in R.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(stringr)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(SRAdb)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">library</span>(Biostrings)</a></code></pre></div>
<p>First, download the database file to our working directly from the following link: <a href="http://dl.dropbox.com/u/51653511/SRAmetadb.sqlite.gz" class="uri">http://dl.dropbox.com/u/51653511/SRAmetadb.sqlite.gz</a> . Note that it’s over 1 GB compressed, but will be about 24 GBs once you uncompress it, which you need to do.</p>
<p>Then, we’ll make a connection with the database:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">sqlfile &lt;-<span class="st"> &#39;~/SRAmetadb.sqlite&#39;</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">sra_con &lt;-<span class="st"> </span><span class="kw">dbConnect</span>(<span class="kw">SQLite</span>(),sqlfile)</a></code></pre></div>
<p>Let’s assume we want to download the data from the Gevers IBD study that was deposited with the following accession: PRJNA237362. We can see this project on NCBI here: <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA237362" class="uri">https://www.ncbi.nlm.nih.gov/bioproject/PRJNA237362</a> . If we click the link next to ‘SRA Experiments’ and then click one of the sample links on the subsequent page, we’ll end up on a page that looks like this: <a href="https://www.ncbi.nlm.nih.gov/sra/SRX1418176%5Baccn%5D" class="uri">https://www.ncbi.nlm.nih.gov/sra/SRX1418176[accn]</a> . Here’ we can access the SRA project code: SRP040765. This is what we need.</p>
<p>Before we continue, let me go over the SRA file types:</p>
<ul>
<li>SRA - Accession information that contains the 5 files below</li>
<li>SRP - Project information and metadata</li>
<li>SRS - Sample metadata</li>
<li>SRX - Experiment metadata including library, platform selection, and processing parametes involved in a particular sequencing experiment</li>
<li>SRR - Sequencing run information</li>
<li>SRX - Sequence analysis BAM file information</li>
</ul>
<p>Now that we have the SRP, let’s acquire the files we need, specifically the SRA files:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">rs &lt;-<span class="st"> </span><span class="kw">listSRAfile</span>(<span class="kw">c</span>(<span class="st">&#39;SRP040765&#39;</span>), sra_con, <span class="dt">fileType =</span> <span class="st">&#39;sra&#39;</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">str</span>(rs)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    3408 obs. of  5 variables:
##  $ study     : chr  &quot;SRP040765&quot; &quot;SRP040765&quot; &quot;SRP040765&quot; &quot;SRP040765&quot; ...
##  $ sample    : chr  &quot;SRS587325&quot; &quot;SRS587957&quot; &quot;SRS587956&quot; &quot;SRS695027&quot; ...
##  $ experiment: chr  &quot;SRX691644&quot; &quot;SRX692390&quot; &quot;SRX509287&quot; &quot;SRX693364&quot; ...
##  $ run       : chr  &quot;SRR1564534&quot; &quot;SRR1565230&quot; &quot;SRR1215330&quot; &quot;SRR1566410&quot; ...
##  $ ftp       : chr  &quot;ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR156/SRR1564534/SRR1564534.sra&quot; &quot;ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR156/SRR1565230/SRR1565230.sra&quot; &quot;ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR121/SRR1215330/SRR1215330.sra&quot; &quot;ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR156/SRR1566410/SRR1566410.sra&quot; ...</code></pre>
<p><strong>rs</strong> is a dataframe containing the SRP, SRS, SRX, and SRR IDs for a given sequencing run, as well as the ftp link to the actual .sra file containing the sequencing information. These are the links we can use to download the entire set of data we need to perform some analysis.</p>
<p>Now, we could export these links and then just iterate through, maybe using bash with wget, to download all of these files. Alternatively, we can do the following:</p>
<p>We’ll get a run ID for a run we’ld like to download:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">run &lt;-<span class="st"> </span>rs<span class="op">$</span>run[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">run</a></code></pre></div>
<pre><code>## [1] &quot;SRR1564534&quot;</code></pre>
<p>If we want the specific SRR (run) information, we do:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">run_info &lt;-<span class="st"> </span><span class="kw">getSRA</span>(<span class="dt">search_terms=</span><span class="st">&#39;SRP040765&#39;</span>, <span class="dt">out_types=</span><span class="kw">c</span>(<span class="st">&#39;run&#39;</span>),sra_con)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw">str</span>(run_info)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    3408 obs. of  11 variables:
##  $ run_alias      : chr  &quot;A21YN121012.1.Illumina_P7-Fexexoja.screened.bam&quot; &quot;A21YN121012.1.Illumina_P7-Wodejora.screened.bam&quot; &quot;A2WP7130403.1.Illumina_P7-Xakokoxe.screened.bam&quot; &quot;A1UEN121219.1.Illumina_P7-Birarane.screened.bam&quot; ...
##  $ run            : chr  &quot;SRR1564534&quot; &quot;SRR1565230&quot; &quot;SRR1215330&quot; &quot;SRR1566410&quot; ...
##  $ run_date       : chr  &quot;2012-10-12&quot; &quot;2012-10-12&quot; &quot;2013-04-03&quot; &quot;2012-12-19&quot; ...
##  $ updated_date   : chr  &quot;2014-09-04&quot; &quot;2014-09-05&quot; &quot;2019-12-11&quot; &quot;2014-09-06&quot; ...
##  $ spots          : num  120 449 2102 9532 676152 ...
##  $ bases          : num  4.20e+04 1.57e+05 7.36e+05 3.34e+06 1.37e+08 ...
##  $ run_center     : chr  &quot;BI&quot; &quot;BI&quot; &quot;BI&quot; &quot;BI&quot; ...
##  $ experiment_name: logi  NA NA NA NA NA NA ...
##  $ run_url_link   : logi  NA NA NA NA NA NA ...
##  $ run_entrez_link: logi  NA NA NA NA NA NA ...
##  $ run_attribute  : chr  &quot;analysis_type: AssemblyWithoutReference || flowcell_barcode: A21YN || gssr_id: 247611.0 || instrument_name: SL-&quot;| __truncated__ &quot;analysis_type: AssemblyWithoutReference || flowcell_barcode: A21YN || gssr_id: 247629.0 || instrument_name: SL-&quot;| __truncated__ &quot;analysis_type: AssemblyWithoutReference || flowcell_barcode: A2WP7 || gssr_id: 247671.0 || instrument_name: SL-&quot;| __truncated__ &quot;analysis_type: AssemblyWithoutReference || data_type: 16S || flowcell_barcode: A1UEN || gssr_id: 276389.0 || in&quot;| __truncated__ ...</code></pre>
<p>and for the SRS (sample) information:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">sample_info &lt;-<span class="st"> </span><span class="kw">getSRA</span>(<span class="dt">search_terms=</span><span class="st">&#39;SRP040765&#39;</span>, <span class="dt">out_types=</span><span class="kw">c</span>(<span class="st">&#39;sample&#39;</span>),sra_con)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw">str</span>(sample_info)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    1572 obs. of  10 variables:
##  $ sample_alias      : chr  &quot;SKBTI-0107&quot; &quot;SKBTI-0128&quot; &quot;SKBTI-0172&quot; &quot;SKBTI-0594&quot; ...
##  $ sample            : chr  &quot;SRS587325&quot; &quot;SRS587957&quot; &quot;SRS587956&quot; &quot;SRS695027&quot; ...
##  $ taxon_id          : int  408170 408170 408170 408170 408170 408170 408170 408170 408170 408170 ...
##  $ common_name       : chr  NA NA NA NA ...
##  $ anonymized_name   : chr  NA NA NA NA ...
##  $ individual_name   : chr  NA NA NA NA ...
##  $ description       : logi  NA NA NA NA NA NA ...
##  $ sample_url_link   : logi  NA NA NA NA NA NA ...
##  $ sample_entrez_link: logi  NA NA NA NA NA NA ...
##  $ sample_attribute  : chr  &quot;strain: SKBTI-0107 || collection date: missing || geographic location (country and/or sea, region): USA || spec&quot;| __truncated__ &quot;strain: SKBTI-0128 || collection date: missing || geographic location (country and/or sea, region): USA || spec&quot;| __truncated__ &quot;strain: SKBTI-0172 || collection date: missing || geographic location (country and/or sea, region): USA || spec&quot;| __truncated__ &quot;strain: SKBTI-0594 || collection date: missing || geographic location (country and/or sea, region): USA || spec&quot;| __truncated__ ...</code></pre>
<p>and SRX (experiment) information:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">experiment_info &lt;-<span class="st"> </span><span class="kw">getSRA</span>(<span class="dt">search_terms=</span><span class="st">&#39;SRP040765&#39;</span>, <span class="dt">out_types=</span><span class="kw">c</span>(<span class="st">&#39;experiment&#39;</span>),sra_con)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw">str</span>(experiment_info)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    2708 obs. of  27 variables:
##  $ experiment_alias             : chr  &quot;2949006.WR32770.Solexa-122962.A21YN121012.P&quot; &quot;2949006.WR32770.Solexa-122980.A21YN121012.P&quot; &quot;2949006.WR32770.Solexa-123022.A2WP7130403.P&quot; &quot;2949006.WR33991.Solexa-133729.A1UEN121219.P&quot; ...
##  $ experiment                   : chr  &quot;SRX691644&quot; &quot;SRX692390&quot; &quot;SRX509287&quot; &quot;SRX693364&quot; ...
##  $ experiment_title             : chr  &quot;Illumina amplicon sequencing of metagenomic paired-end library &#39;Solexa-122962&#39; containing sample &#39;SKBTI-0107&#39;&quot; &quot;Illumina amplicon sequencing of metagenomic paired-end library &#39;Solexa-122980&#39; containing sample &#39;SKBTI-0128&#39;&quot; &quot;Illumina amplicon sequencing of metagenomic paired-end library &#39;Solexa-123022&#39; containing sample &#39;SKBTI-0172&#39;&quot; &quot;Illumina amplicon sequencing of metagenomic paired-end library &#39;Solexa-133729&#39; containing sample &#39;SKBTI-0594&#39;&quot; ...
##  $ study_name                   : logi  NA NA NA NA NA NA ...
##  $ sample_name                  : logi  NA NA NA NA NA NA ...
##  $ design_description           : chr  &quot;Illumina sequencing of human gut metagenome via polymerase chain reaction&quot; &quot;Illumina sequencing of human gut metagenome via polymerase chain reaction&quot; &quot;Illumina sequencing of human gut metagenome via polymerase chain reaction&quot; &quot;Illumina sequencing of human gut metagenome via polymerase chain reaction&quot; ...
##  $ library_name                 : chr  &quot;Solexa-122962&quot; &quot;Solexa-122980&quot; &quot;Solexa-123022&quot; &quot;Solexa-133729&quot; ...
##  $ library_strategy             : chr  &quot;AMPLICON&quot; &quot;AMPLICON&quot; &quot;AMPLICON&quot; &quot;AMPLICON&quot; ...
##  $ library_source               : chr  &quot;METAGENOMIC&quot; &quot;METAGENOMIC&quot; &quot;METAGENOMIC&quot; &quot;METAGENOMIC&quot; ...
##  $ library_selection            : chr  &quot;PCR&quot; &quot;PCR&quot; &quot;PCR&quot; &quot;PCR&quot; ...
##  $ library_layout               : chr  &quot;PAIRED - NOMINAL_SDEV: 0.0E0; NOMINAL_LENGTH: 390; &quot; &quot;PAIRED - NOMINAL_SDEV: 0.0E0; NOMINAL_LENGTH: 390; &quot; &quot;PAIRED - NOMINAL_SDEV: 0.0E0; NOMINAL_LENGTH: 393; &quot; &quot;PAIRED - NOMINAL_SDEV: 0.0E0; NOMINAL_LENGTH: 382; &quot; ...
##  $ library_construction_protocol: logi  NA NA NA NA NA NA ...
##  $ adapter_spec                 : logi  NA NA NA NA NA NA ...
##  $ read_spec                    : chr  &quot;READ_INDEX: 0; READ_LABEL: forward; READ_CLASS: Application Read; READ_TYPE: Forward; BASE_COORD: 1 || READ_IND&quot;| __truncated__ &quot;READ_INDEX: 0; READ_LABEL: forward; READ_CLASS: Application Read; READ_TYPE: Forward; BASE_COORD: 1 || READ_IND&quot;| __truncated__ &quot;READ_INDEX: 0; READ_LABEL: forward; READ_CLASS: Application Read; READ_TYPE: Forward; BASE_COORD: 1 || READ_IND&quot;| __truncated__ &quot;READ_INDEX: 0; READ_LABEL: forward; READ_CLASS: Application Read; READ_TYPE: Forward; BASE_COORD: 1 || READ_IND&quot;| __truncated__ ...
##  $ platform                     : chr  &quot;ILLUMINA&quot; &quot;ILLUMINA&quot; &quot;ILLUMINA&quot; &quot;ILLUMINA&quot; ...
##  $ instrument_model             : chr  &quot;Illumina MiSeq&quot; &quot;Illumina MiSeq&quot; &quot;Illumina MiSeq&quot; &quot;Illumina MiSeq&quot; ...
##  $ instrument_name              : logi  NA NA NA NA NA NA ...
##  $ platform_parameters          : chr  &quot;INSTRUMENT_MODEL: Illumina MiSeq&quot; &quot;INSTRUMENT_MODEL: Illumina MiSeq&quot; &quot;INSTRUMENT_MODEL: Illumina MiSeq&quot; &quot;INSTRUMENT_MODEL: Illumina MiSeq&quot; ...
##  $ sequence_space               : logi  NA NA NA NA NA NA ...
##  $ base_caller                  : logi  NA NA NA NA NA NA ...
##  $ quality_scorer               : logi  NA NA NA NA NA NA ...
##  $ number_of_levels             : logi  NA NA NA NA NA NA ...
##  $ multiplier                   : logi  NA NA NA NA NA NA ...
##  $ qtype                        : logi  NA NA NA NA NA NA ...
##  $ experiment_url_link          : logi  NA NA NA NA NA NA ...
##  $ experiment_entrez_link       : logi  NA NA NA NA NA NA ...
##  $ experiment_attribute         : chr  &quot;analysis_type: AssemblyWithoutReference || gssr_id: 247611.0 || library_type: 16S || lsid: broadinstitute.org:b&quot;| __truncated__ &quot;analysis_type: AssemblyWithoutReference || gssr_id: 247629.0 || library_type: 16S || lsid: broadinstitute.org:b&quot;| __truncated__ &quot;analysis_type: AssemblyWithoutReference || gssr_id: 247671.0 || library_type: 16S || lsid: broadinstitute.org:b&quot;| __truncated__ &quot;analysis_type: AssemblyWithoutReference || data_type: 16S || gssr_id: 276389.0 || library_type: 16S || lsid: br&quot;| __truncated__ ...</code></pre>
<p>Using these commands, you should be able to download the .sra files you need along with all corresponding metadata to do analysis. Still, you might be wondering how you get the .fasta files from the .sra file. Well, the easiest way is to use something called the sra toolkit, which can be found here: <a href="https://www.ncbi.nlm.nih.gov/books/NBK158900/" class="uri">https://www.ncbi.nlm.nih.gov/books/NBK158900/</a> .</p>
<p>So let’s say we aimed to extract the sequences from the following sra files: SRR1635768 and SRR1566401. First, we’d download the files:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">sra_dir &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"></a>
<a class="sourceLine" id="cb13-3" data-line-number="3">sra_fns &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;SRR1634425&quot;</span>,<span class="st">&quot;SRR1634428&quot;</span>)</a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="cf">for</span> (sra <span class="cf">in</span> sra_fns) <span class="kw">getSRAfile</span>(sra, sra_con, <span class="dt">fileType =</span> <span class="st">&#39;sra&#39;</span>,<span class="dt">destDir=</span>sra_dir)</a></code></pre></div>
<pre><code>## Files are saved to: 
## &#39;/tmp/RtmpF5vs4m&#39;
## 
## Files are saved to: 
## &#39;/tmp/RtmpF5vs4m&#39;</code></pre>
<p>Then, we’d use the sra toolkit to extract the sequences. Assuming you have downloaded and installed it, we can do the following</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">sra_output &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"></a>
<a class="sourceLine" id="cb15-3" data-line-number="3">sra_files &lt;-<span class="st"> </span><span class="kw">list.files</span>(sra_dir,<span class="dt">full.names=</span><span class="ot">TRUE</span>,<span class="dt">pattern=</span><span class="st">&#39;</span><span class="ch">\\</span><span class="st">.sra$&#39;</span>)</a>
<a class="sourceLine" id="cb15-4" data-line-number="4"></a>
<a class="sourceLine" id="cb15-5" data-line-number="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(sra_files)) <span class="kw">system2</span>(<span class="st">&#39;fastq-dump&#39;</span>,<span class="dt">args=</span><span class="kw">c</span>(sra_files[i],</a>
<a class="sourceLine" id="cb15-6" data-line-number="6">                                                              <span class="st">&#39;-O&#39;</span>, sra_output,</a>
<a class="sourceLine" id="cb15-7" data-line-number="7">                                                              <span class="st">&#39;--gzip&#39;</span>,</a>
<a class="sourceLine" id="cb15-8" data-line-number="8">                                                              <span class="st">&#39;--clip&#39;</span>,</a>
<a class="sourceLine" id="cb15-9" data-line-number="9">                                                              <span class="st">&#39;--skip-technical&#39;</span>,</a>
<a class="sourceLine" id="cb15-10" data-line-number="10">                                                              <span class="st">&#39;--dumpbase&#39;</span>))</a></code></pre></div>
<p>And now we can check:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">fqs &lt;-<span class="st"> </span><span class="kw">list.files</span>(sra_output,<span class="dt">full.names=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"></a>
<a class="sourceLine" id="cb16-3" data-line-number="3">FASTQ &lt;-<span class="st"> </span><span class="kw">readDNAStringSet</span>(fqs,<span class="dt">format=</span><span class="st">&#39;fastq&#39;</span>)</a>
<a class="sourceLine" id="cb16-4" data-line-number="4">FASTQ</a></code></pre></div>
<p>Note the header names; they’re simply the SRR ids. You’d have to the sra metadata (see above) to match these to specific samples. For example,</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">sample_info &lt;-<span class="st"> </span><span class="kw">getSRA</span>(<span class="dt">search_terms=</span><span class="st">&#39;SRR1635768&#39;</span>, <span class="dt">out_types=</span><span class="kw">c</span>(<span class="st">&#39;sample&#39;</span>),sra_con)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw">str</span>(sample_info)</a></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    1 obs. of  10 variables:
##  $ sample_alias      : chr &quot;SKBTI-0325&quot;
##  $ sample            : chr &quot;SRS734393&quot;
##  $ taxon_id          : int 408170
##  $ common_name       : logi NA
##  $ anonymized_name   : logi NA
##  $ individual_name   : logi NA
##  $ description       : logi NA
##  $ sample_url_link   : logi NA
##  $ sample_entrez_link: logi NA
##  $ sample_attribute  : chr &quot;strain: SKBTI-0325 || collection date: missing || geographic location (country and/or sea, region): USA || spec&quot;| __truncated__</code></pre>
<div id="fastq-dump-for-paired-end-reads" class="section level3">
<h3><span class="header-section-number">6.0.1</span> Fastq Dump for Paired End Reads</h3>
<p>It should be noted that simply using the fastq-dump command as above works only if your data <em>does not consist of paired end reads</em>. If you happen to have paired end reads, then the following arguments must be added to the fastq-dump call:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(sra_files)) <span class="kw">system2</span>(<span class="st">&#39;fastq-dump&#39;</span>,<span class="dt">args=</span><span class="kw">c</span>(sra_files[i],</a>
<a class="sourceLine" id="cb19-2" data-line-number="2">                                                              <span class="st">&#39;-O&#39;</span>, sra_output,</a>
<a class="sourceLine" id="cb19-3" data-line-number="3">                                                              <span class="st">&#39;--gzip&#39;</span>,</a>
<a class="sourceLine" id="cb19-4" data-line-number="4">                                                              <span class="st">&#39;--clip&#39;</span>,</a>
<a class="sourceLine" id="cb19-5" data-line-number="5">                                                              <span class="st">&#39;--skip-technical&#39;</span>,</a>
<a class="sourceLine" id="cb19-6" data-line-number="6">                                                              <span class="st">&#39;--dumpbase&#39;</span>,</a>
<a class="sourceLine" id="cb19-7" data-line-number="7">                                                              <span class="st">&#39;--split-files&#39;</span>,</a>
<a class="sourceLine" id="cb19-8" data-line-number="8">                                                              <span class="st">&#39;--readids&#39;</span>))</a></code></pre></div>
<p>For more information, see <a href="https://edwards.sdsu.edu/research/fastq-dump/" class="uri">https://edwards.sdsu.edu/research/fastq-dump/</a></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="biocond.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="phyloseq.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bioinformatics_class.pdf", "bioinformatics_class.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
